// thirdweb/dist/cjs/extensions/erc1155/__generated__/INFTMetadata/read/supportsInterface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.supportsInterface = exports.decodeSupportsInterfaceResult = exports.encodeSupportsInterface = exports.encodeSupportsInterfaceParams = exports.isSupportsInterfaceSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x01ffc9a7";
const FN_INPUTS = [
    {
        type: "bytes4",
        name: "interfaceId",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `supportsInterface` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `supportsInterface` method is supported.
 * @extension ERC1155
 * @example
 * ```ts
 * import { isSupportsInterfaceSupported } from "thirdweb/extensions/erc1155";
 *
 * const supported = await isSupportsInterfaceSupported(contract);
 * ```
 */
async function isSupportsInterfaceSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isSupportsInterfaceSupported = isSupportsInterfaceSupported;
/**
 * Encodes the parameters for the "supportsInterface" function.
 * @param options - The options for the supportsInterface function.
 * @returns The encoded ABI parameters.
 * @extension ERC1155
 * @example
 * ```ts
 * import { encodeSupportsInterfaceParams } "thirdweb/extensions/erc1155";
 * const result = encodeSupportsInterfaceParams({
 *  interfaceId: ...,
 * });
 * ```
 */
function encodeSupportsInterfaceParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.interfaceId]);
}
exports.encodeSupportsInterfaceParams = encodeSupportsInterfaceParams;
/**
 * Encodes the "supportsInterface" function into a Hex string with its parameters.
 * @param options - The options for the supportsInterface function.
 * @returns The encoded hexadecimal string.
 * @extension ERC1155
 * @example
 * ```ts
 * import { encodeSupportsInterface } "thirdweb/extensions/erc1155";
 * const result = encodeSupportsInterface({
 *  interfaceId: ...,
 * });
 * ```
 */
function encodeSupportsInterface(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeSupportsInterfaceParams(options).slice(2));
}
exports.encodeSupportsInterface = encodeSupportsInterface;
/**
 * Decodes the result of the supportsInterface function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC1155
 * @example
 * ```ts
 * import { decodeSupportsInterfaceResult } from "thirdweb/extensions/erc1155";
 * const result = decodeSupportsInterfaceResult("...");
 * ```
 */
function decodeSupportsInterfaceResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeSupportsInterfaceResult = decodeSupportsInterfaceResult;
/**
 * Calls the "supportsInterface" function on the contract.
 * @param options - The options for the supportsInterface function.
 * @returns The parsed result of the function call.
 * @extension ERC1155
 * @example
 * ```ts
 * import { supportsInterface } from "thirdweb/extensions/erc1155";
 *
 * const result = await supportsInterface({
 *  contract,
 *  interfaceId: ...,
 * });
 *
 * ```
 */
async function supportsInterface(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.interfaceId],
    });
}
exports.supportsInterface = supportsInterface;
