// thirdweb/dist/esm/extensions/erc20/__generated__/IStaking20/write/claimRewards.js
import { prepareContractCall } from "../../../../../transaction/prepare-contract-call.js";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x372500ab";
const FN_INPUTS = [];
const FN_OUTPUTS = [];
/**
 * Checks if the `claimRewards` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `claimRewards` method is supported.
 * @extension ERC20
 * @example
 * ```ts
 * import { isClaimRewardsSupported } from "thirdweb/extensions/erc20";
 *
 * const supported = await isClaimRewardsSupported(contract);
 * ```
 */
export async function isClaimRewardsSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Prepares a transaction to call the "claimRewards" function on the contract.
 * @param options - The options for the "claimRewards" function.
 * @returns A prepared transaction object.
 * @extension ERC20
 * @example
 * ```ts
 * import { claimRewards } from "thirdweb/extensions/erc20";
 *
 * const transaction = claimRewards();
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
export function claimRewards(options) {
    return prepareContractCall({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
