// thirdweb/dist/cjs/extensions/erc721/__generated__/IERC721Receiver/write/onERC721Received.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.onERC721Received = exports.encodeOnERC721Received = exports.encodeOnERC721ReceivedParams = exports.isOnERC721ReceivedSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x150b7a02";
const FN_INPUTS = [
    {
        type: "address",
        name: "operator",
    },
    {
        type: "address",
        name: "from",
    },
    {
        type: "uint256",
        name: "tokenId",
    },
    {
        type: "bytes",
        name: "data",
    },
];
const FN_OUTPUTS = [
    {
        type: "bytes4",
    },
];
/**
 * Checks if the `onERC721Received` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `onERC721Received` method is supported.
 * @extension ERC721
 * @example
 * ```ts
 * import { isOnERC721ReceivedSupported } from "thirdweb/extensions/erc721";
 *
 * const supported = await isOnERC721ReceivedSupported(contract);
 * ```
 */
async function isOnERC721ReceivedSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isOnERC721ReceivedSupported = isOnERC721ReceivedSupported;
/**
 * Encodes the parameters for the "onERC721Received" function.
 * @param options - The options for the onERC721Received function.
 * @returns The encoded ABI parameters.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeOnERC721ReceivedParams } "thirdweb/extensions/erc721";
 * const result = encodeOnERC721ReceivedParams({
 *  operator: ...,
 *  from: ...,
 *  tokenId: ...,
 *  data: ...,
 * });
 * ```
 */
function encodeOnERC721ReceivedParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [
        options.operator,
        options.from,
        options.tokenId,
        options.data,
    ]);
}
exports.encodeOnERC721ReceivedParams = encodeOnERC721ReceivedParams;
/**
 * Encodes the "onERC721Received" function into a Hex string with its parameters.
 * @param options - The options for the onERC721Received function.
 * @returns The encoded hexadecimal string.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeOnERC721Received } "thirdweb/extensions/erc721";
 * const result = encodeOnERC721Received({
 *  operator: ...,
 *  from: ...,
 *  tokenId: ...,
 *  data: ...,
 * });
 * ```
 */
function encodeOnERC721Received(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeOnERC721ReceivedParams(options).slice(2));
}
exports.encodeOnERC721Received = encodeOnERC721Received;
/**
 * Prepares a transaction to call the "onERC721Received" function on the contract.
 * @param options - The options for the "onERC721Received" function.
 * @returns A prepared transaction object.
 * @extension ERC721
 * @example
 * ```ts
 * import { onERC721Received } from "thirdweb/extensions/erc721";
 *
 * const transaction = onERC721Received({
 *  contract,
 *  operator: ...,
 *  from: ...,
 *  tokenId: ...,
 *  data: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function onERC721Received(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [
                resolvedOptions.operator,
                resolvedOptions.from,
                resolvedOptions.tokenId,
                resolvedOptions.data,
            ];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.onERC721Received = onERC721Received;
