// thirdweb/dist/cjs/extensions/erc20/__generated__/IERC20/read/balanceOf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.balanceOf = exports.decodeBalanceOfResult = exports.encodeBalanceOf = exports.encodeBalanceOfParams = exports.isBalanceOfSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x70a08231";
const FN_INPUTS = [
    {
        type: "address",
        name: "_address",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `balanceOf` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `balanceOf` method is supported.
 * @extension ERC20
 * @example
 * ```ts
 * import { isBalanceOfSupported } from "thirdweb/extensions/erc20";
 *
 * const supported = await isBalanceOfSupported(contract);
 * ```
 */
async function isBalanceOfSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isBalanceOfSupported = isBalanceOfSupported;
/**
 * Encodes the parameters for the "balanceOf" function.
 * @param options - The options for the balanceOf function.
 * @returns The encoded ABI parameters.
 * @extension ERC20
 * @example
 * ```ts
 * import { encodeBalanceOfParams } "thirdweb/extensions/erc20";
 * const result = encodeBalanceOfParams({
 *  address: ...,
 * });
 * ```
 */
function encodeBalanceOfParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.address]);
}
exports.encodeBalanceOfParams = encodeBalanceOfParams;
/**
 * Encodes the "balanceOf" function into a Hex string with its parameters.
 * @param options - The options for the balanceOf function.
 * @returns The encoded hexadecimal string.
 * @extension ERC20
 * @example
 * ```ts
 * import { encodeBalanceOf } "thirdweb/extensions/erc20";
 * const result = encodeBalanceOf({
 *  address: ...,
 * });
 * ```
 */
function encodeBalanceOf(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeBalanceOfParams(options).slice(2));
}
exports.encodeBalanceOf = encodeBalanceOf;
/**
 * Decodes the result of the balanceOf function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC20
 * @example
 * ```ts
 * import { decodeBalanceOfResult } from "thirdweb/extensions/erc20";
 * const result = decodeBalanceOfResult("...");
 * ```
 */
function decodeBalanceOfResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeBalanceOfResult = decodeBalanceOfResult;
/**
 * Calls the "balanceOf" function on the contract.
 * @param options - The options for the balanceOf function.
 * @returns The parsed result of the function call.
 * @extension ERC20
 * @example
 * ```ts
 * import { balanceOf } from "thirdweb/extensions/erc20";
 *
 * const result = await balanceOf({
 *  contract,
 *  address: ...,
 * });
 *
 * ```
 */
async function balanceOf(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.address],
    });
}
exports.balanceOf = balanceOf;
