// thirdweb/dist/cjs/extensions/erc20/__generated__/IERC20/read/totalSupply.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.totalSupply = exports.decodeTotalSupplyResult = exports.isTotalSupplySupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x18160ddd";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `totalSupply` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `totalSupply` method is supported.
 * @extension ERC20
 * @example
 * ```ts
 * import { isTotalSupplySupported } from "thirdweb/extensions/erc20";
 *
 * const supported = await isTotalSupplySupported(contract);
 * ```
 */
async function isTotalSupplySupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isTotalSupplySupported = isTotalSupplySupported;
/**
 * Decodes the result of the totalSupply function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC20
 * @example
 * ```ts
 * import { decodeTotalSupplyResult } from "thirdweb/extensions/erc20";
 * const result = decodeTotalSupplyResult("...");
 * ```
 */
function decodeTotalSupplyResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeTotalSupplyResult = decodeTotalSupplyResult;
/**
 * Calls the "totalSupply" function on the contract.
 * @param options - The options for the totalSupply function.
 * @returns The parsed result of the function call.
 * @extension ERC20
 * @example
 * ```ts
 * import { totalSupply } from "thirdweb/extensions/erc20";
 *
 * const result = await totalSupply({
 *  contract,
 * });
 *
 * ```
 */
async function totalSupply(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.totalSupply = totalSupply;
