// thirdweb/dist/cjs/extensions/farcaster/__generated__/IIdRegistry/write/transferAndChangeRecoveryFor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferAndChangeRecoveryFor = exports.encodeTransferAndChangeRecoveryFor = exports.encodeTransferAndChangeRecoveryForParams = exports.isTransferAndChangeRecoveryForSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x4c5cbb34";
const FN_INPUTS = [
    {
        type: "address",
        name: "from",
    },
    {
        type: "address",
        name: "to",
    },
    {
        type: "address",
        name: "recovery",
    },
    {
        type: "uint256",
        name: "fromDeadline",
    },
    {
        type: "bytes",
        name: "fromSig",
    },
    {
        type: "uint256",
        name: "toDeadline",
    },
    {
        type: "bytes",
        name: "toSig",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `transferAndChangeRecoveryFor` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `transferAndChangeRecoveryFor` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isTransferAndChangeRecoveryForSupported } from "thirdweb/extensions/farcaster";
 *
 * const supported = await isTransferAndChangeRecoveryForSupported(contract);
 * ```
 */
async function isTransferAndChangeRecoveryForSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isTransferAndChangeRecoveryForSupported = isTransferAndChangeRecoveryForSupported;
/**
 * Encodes the parameters for the "transferAndChangeRecoveryFor" function.
 * @param options - The options for the transferAndChangeRecoveryFor function.
 * @returns The encoded ABI parameters.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { encodeTransferAndChangeRecoveryForParams } "thirdweb/extensions/farcaster";
 * const result = encodeTransferAndChangeRecoveryForParams({
 *  from: ...,
 *  to: ...,
 *  recovery: ...,
 *  fromDeadline: ...,
 *  fromSig: ...,
 *  toDeadline: ...,
 *  toSig: ...,
 * });
 * ```
 */
function encodeTransferAndChangeRecoveryForParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [
        options.from,
        options.to,
        options.recovery,
        options.fromDeadline,
        options.fromSig,
        options.toDeadline,
        options.toSig,
    ]);
}
exports.encodeTransferAndChangeRecoveryForParams = encodeTransferAndChangeRecoveryForParams;
/**
 * Encodes the "transferAndChangeRecoveryFor" function into a Hex string with its parameters.
 * @param options - The options for the transferAndChangeRecoveryFor function.
 * @returns The encoded hexadecimal string.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { encodeTransferAndChangeRecoveryFor } "thirdweb/extensions/farcaster";
 * const result = encodeTransferAndChangeRecoveryFor({
 *  from: ...,
 *  to: ...,
 *  recovery: ...,
 *  fromDeadline: ...,
 *  fromSig: ...,
 *  toDeadline: ...,
 *  toSig: ...,
 * });
 * ```
 */
function encodeTransferAndChangeRecoveryFor(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeTransferAndChangeRecoveryForParams(options).slice(2));
}
exports.encodeTransferAndChangeRecoveryFor = encodeTransferAndChangeRecoveryFor;
/**
 * Prepares a transaction to call the "transferAndChangeRecoveryFor" function on the contract.
 * @param options - The options for the "transferAndChangeRecoveryFor" function.
 * @returns A prepared transaction object.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { transferAndChangeRecoveryFor } from "thirdweb/extensions/farcaster";
 *
 * const transaction = transferAndChangeRecoveryFor({
 *  contract,
 *  from: ...,
 *  to: ...,
 *  recovery: ...,
 *  fromDeadline: ...,
 *  fromSig: ...,
 *  toDeadline: ...,
 *  toSig: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function transferAndChangeRecoveryFor(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [
                resolvedOptions.from,
                resolvedOptions.to,
                resolvedOptions.recovery,
                resolvedOptions.fromDeadline,
                resolvedOptions.fromSig,
                resolvedOptions.toDeadline,
                resolvedOptions.toSig,
            ];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.transferAndChangeRecoveryFor = transferAndChangeRecoveryFor;
