// thirdweb/dist/cjs/extensions/marketplace/__generated__/IDirectListings/read/currencyPriceForListing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.currencyPriceForListing = exports.decodeCurrencyPriceForListingResult = exports.encodeCurrencyPriceForListing = exports.encodeCurrencyPriceForListingParams = exports.isCurrencyPriceForListingSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xfb14079d";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_listingId",
    },
    {
        type: "address",
        name: "_currency",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `currencyPriceForListing` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `currencyPriceForListing` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isCurrencyPriceForListingSupported } from "thirdweb/extensions/marketplace";
 *
 * const supported = await isCurrencyPriceForListingSupported(contract);
 * ```
 */
async function isCurrencyPriceForListingSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isCurrencyPriceForListingSupported = isCurrencyPriceForListingSupported;
/**
 * Encodes the parameters for the "currencyPriceForListing" function.
 * @param options - The options for the currencyPriceForListing function.
 * @returns The encoded ABI parameters.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { encodeCurrencyPriceForListingParams } "thirdweb/extensions/marketplace";
 * const result = encodeCurrencyPriceForListingParams({
 *  listingId: ...,
 *  currency: ...,
 * });
 * ```
 */
function encodeCurrencyPriceForListingParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.listingId, options.currency]);
}
exports.encodeCurrencyPriceForListingParams = encodeCurrencyPriceForListingParams;
/**
 * Encodes the "currencyPriceForListing" function into a Hex string with its parameters.
 * @param options - The options for the currencyPriceForListing function.
 * @returns The encoded hexadecimal string.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { encodeCurrencyPriceForListing } "thirdweb/extensions/marketplace";
 * const result = encodeCurrencyPriceForListing({
 *  listingId: ...,
 *  currency: ...,
 * });
 * ```
 */
function encodeCurrencyPriceForListing(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeCurrencyPriceForListingParams(options).slice(2));
}
exports.encodeCurrencyPriceForListing = encodeCurrencyPriceForListing;
/**
 * Decodes the result of the currencyPriceForListing function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { decodeCurrencyPriceForListingResult } from "thirdweb/extensions/marketplace";
 * const result = decodeCurrencyPriceForListingResult("...");
 * ```
 */
function decodeCurrencyPriceForListingResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeCurrencyPriceForListingResult = decodeCurrencyPriceForListingResult;
/**
 * Calls the "currencyPriceForListing" function on the contract.
 * @param options - The options for the currencyPriceForListing function.
 * @returns The parsed result of the function call.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { currencyPriceForListing } from "thirdweb/extensions/marketplace";
 *
 * const result = await currencyPriceForListing({
 *  contract,
 *  listingId: ...,
 *  currency: ...,
 * });
 *
 * ```
 */
async function currencyPriceForListing(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.listingId, options.currency],
    });
}
exports.currencyPriceForListing = currencyPriceForListing;
