// thirdweb/dist/cjs/extensions/marketplace/__generated__/IDirectListings/read/totalListings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.totalListings = exports.decodeTotalListingsResult = exports.isTotalListingsSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xc78b616c";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `totalListings` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `totalListings` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isTotalListingsSupported } from "thirdweb/extensions/marketplace";
 *
 * const supported = await isTotalListingsSupported(contract);
 * ```
 */
async function isTotalListingsSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isTotalListingsSupported = isTotalListingsSupported;
/**
 * Decodes the result of the totalListings function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { decodeTotalListingsResult } from "thirdweb/extensions/marketplace";
 * const result = decodeTotalListingsResult("...");
 * ```
 */
function decodeTotalListingsResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeTotalListingsResult = decodeTotalListingsResult;
/**
 * Calls the "totalListings" function on the contract.
 * @param options - The options for the totalListings function.
 * @returns The parsed result of the function call.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { totalListings } from "thirdweb/extensions/marketplace";
 *
 * const result = await totalListings({
 *  contract,
 * });
 *
 * ```
 */
async function totalListings(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.totalListings = totalListings;
