// thirdweb/dist/cjs/extensions/erc1155/__generated__/IDrop1155/read/getActiveClaimConditionId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getActiveClaimConditionId = exports.decodeGetActiveClaimConditionIdResult = exports.encodeGetActiveClaimConditionId = exports.encodeGetActiveClaimConditionIdParams = exports.isGetActiveClaimConditionIdSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x5ab063e8";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_tokenId",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getActiveClaimConditionId` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getActiveClaimConditionId` method is supported.
 * @extension ERC1155
 * @example
 * ```ts
 * import { isGetActiveClaimConditionIdSupported } from "thirdweb/extensions/erc1155";
 *
 * const supported = await isGetActiveClaimConditionIdSupported(contract);
 * ```
 */
async function isGetActiveClaimConditionIdSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetActiveClaimConditionIdSupported = isGetActiveClaimConditionIdSupported;
/**
 * Encodes the parameters for the "getActiveClaimConditionId" function.
 * @param options - The options for the getActiveClaimConditionId function.
 * @returns The encoded ABI parameters.
 * @extension ERC1155
 * @example
 * ```ts
 * import { encodeGetActiveClaimConditionIdParams } "thirdweb/extensions/erc1155";
 * const result = encodeGetActiveClaimConditionIdParams({
 *  tokenId: ...,
 * });
 * ```
 */
function encodeGetActiveClaimConditionIdParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.tokenId]);
}
exports.encodeGetActiveClaimConditionIdParams = encodeGetActiveClaimConditionIdParams;
/**
 * Encodes the "getActiveClaimConditionId" function into a Hex string with its parameters.
 * @param options - The options for the getActiveClaimConditionId function.
 * @returns The encoded hexadecimal string.
 * @extension ERC1155
 * @example
 * ```ts
 * import { encodeGetActiveClaimConditionId } "thirdweb/extensions/erc1155";
 * const result = encodeGetActiveClaimConditionId({
 *  tokenId: ...,
 * });
 * ```
 */
function encodeGetActiveClaimConditionId(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetActiveClaimConditionIdParams(options).slice(2));
}
exports.encodeGetActiveClaimConditionId = encodeGetActiveClaimConditionId;
/**
 * Decodes the result of the getActiveClaimConditionId function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC1155
 * @example
 * ```ts
 * import { decodeGetActiveClaimConditionIdResult } from "thirdweb/extensions/erc1155";
 * const result = decodeGetActiveClaimConditionIdResult("...");
 * ```
 */
function decodeGetActiveClaimConditionIdResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetActiveClaimConditionIdResult = decodeGetActiveClaimConditionIdResult;
/**
 * Calls the "getActiveClaimConditionId" function on the contract.
 * @param options - The options for the getActiveClaimConditionId function.
 * @returns The parsed result of the function call.
 * @extension ERC1155
 * @example
 * ```ts
 * import { getActiveClaimConditionId } from "thirdweb/extensions/erc1155";
 *
 * const result = await getActiveClaimConditionId({
 *  contract,
 *  tokenId: ...,
 * });
 *
 * ```
 */
async function getActiveClaimConditionId(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.tokenId],
    });
}
exports.getActiveClaimConditionId = getActiveClaimConditionId;
