// thirdweb/dist/cjs/wallets/in-app/implementations/utils/iFrameCommunication/InAppWalletIframeCommunicator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IN_APP_WALLET_IFRAME_ID = exports.createInAppWalletIframeLink = exports.InAppWalletIframeCommunicator = void 0;
const settings_js_1 = require("../../constants/settings.js");
const LocalStorage_js_1 = require("../Storage/LocalStorage.js");
const IframeCommunicator_js_1 = require("./IframeCommunicator.js");
/**
 * @internal
 */
class InAppWalletIframeCommunicator extends IframeCommunicator_js_1.IframeCommunicator {
    /**
     * @internal
     */
    constructor({ clientId, baseUrl }) {
        super({
            iframeId: exports.IN_APP_WALLET_IFRAME_ID,
            link: createInAppWalletIframeLink({
                clientId,
                path: settings_js_1.IN_APP_WALLET_PATH,
                baseUrl,
            }).href,
            baseUrl,
            container: document.body,
        });
        Object.defineProperty(this, "clientId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.clientId = clientId;
    }
    /**
     * @internal
     */
    async onIframeLoadedInitVariables() {
        const localStorage = new LocalStorage_js_1.LocalStorage({
            clientId: this.clientId,
        });
        return {
            authCookie: await localStorage.getAuthCookie(),
            deviceShareStored: await localStorage.getDeviceShare(),
            walletUserId: await localStorage.getWalletUserId(),
            clientId: this.clientId,
        };
    }
}
exports.InAppWalletIframeCommunicator = InAppWalletIframeCommunicator;
// This is the URL and ID tag of the iFrame that we communicate with
/**
 * @internal
 */
function createInAppWalletIframeLink({ clientId, baseUrl, path, queryParams, }) {
    const inAppWalletUrl = new URL(`${path}`, baseUrl);
    if (queryParams) {
        for (const queryKey of Object.keys(queryParams)) {
            inAppWalletUrl.searchParams.set(queryKey, queryParams[queryKey]?.toString() || "");
        }
    }
    inAppWalletUrl.searchParams.set("clientId", clientId);
    return inAppWalletUrl;
}
exports.createInAppWalletIframeLink = createInAppWalletIframeLink;
exports.IN_APP_WALLET_IFRAME_ID = "thirdweb-in-app-wallet-iframe";
