// thirdweb/dist/cjs/extensions/erc1155/drops/write/claimTo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.claimTo = void 0;
const get_claim_params_js_1 = require("../../../../utils/extensions/drops/get-claim-params.js");
const claim_js_1 = require("../../__generated__/IDrop1155/write/claim.js");
/**
 * Claim ERC1155 NFTs to a specified address
 * @param options - The options for the transaction
 * @extension ERC1155
 * @example
 * ```ts
 * import { claimTo } from "thirdweb/extensions/erc1155";
 *
 * const tx = await claimTo({
 *   contract,
 *   to: "0x...",
 *   tokenId: 0n,
 *   quantity: 1n,
 * });
 * ```
 * @throws If no claim condition is set
 * @returns The prepared transaction
 */
function claimTo(options) {
    return (0, claim_js_1.claim)({
        contract: options.contract,
        async asyncParams() {
            const params = await (0, get_claim_params_js_1.getClaimParams)({
                type: "erc1155",
                contract: options.contract,
                to: options.to,
                quantity: options.quantity,
                from: options.from,
                tokenId: options.tokenId,
            });
            return {
                ...params,
                tokenId: options.tokenId,
            };
        },
    });
}
exports.claimTo = claimTo;
