// thirdweb/dist/cjs/extensions/farcaster/__generated__/IBundler/read/keyGateway.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.keyGateway = exports.decodeKeyGatewayResult = exports.isKeyGatewaySupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x80334737";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "address",
    },
];
/**
 * Checks if the `keyGateway` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `keyGateway` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isKeyGatewaySupported } from "thirdweb/extensions/farcaster";
 *
 * const supported = await isKeyGatewaySupported(contract);
 * ```
 */
async function isKeyGatewaySupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isKeyGatewaySupported = isKeyGatewaySupported;
/**
 * Decodes the result of the keyGateway function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { decodeKeyGatewayResult } from "thirdweb/extensions/farcaster";
 * const result = decodeKeyGatewayResult("...");
 * ```
 */
function decodeKeyGatewayResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeKeyGatewayResult = decodeKeyGatewayResult;
/**
 * Calls the "keyGateway" function on the contract.
 * @param options - The options for the keyGateway function.
 * @returns The parsed result of the function call.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { keyGateway } from "thirdweb/extensions/farcaster";
 *
 * const result = await keyGateway({
 *  contract,
 * });
 *
 * ```
 */
async function keyGateway(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.keyGateway = keyGateway;
