// thirdweb/dist/cjs/extensions/farcaster/__generated__/IBundler/read/price.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.price = exports.decodePriceResult = exports.encodePrice = exports.encodePriceParams = exports.isPriceSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x26a49e37";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "extraStorage",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `price` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `price` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isPriceSupported } from "thirdweb/extensions/farcaster";
 *
 * const supported = await isPriceSupported(contract);
 * ```
 */
async function isPriceSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isPriceSupported = isPriceSupported;
/**
 * Encodes the parameters for the "price" function.
 * @param options - The options for the price function.
 * @returns The encoded ABI parameters.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { encodePriceParams } "thirdweb/extensions/farcaster";
 * const result = encodePriceParams({
 *  extraStorage: ...,
 * });
 * ```
 */
function encodePriceParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.extraStorage]);
}
exports.encodePriceParams = encodePriceParams;
/**
 * Encodes the "price" function into a Hex string with its parameters.
 * @param options - The options for the price function.
 * @returns The encoded hexadecimal string.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { encodePrice } "thirdweb/extensions/farcaster";
 * const result = encodePrice({
 *  extraStorage: ...,
 * });
 * ```
 */
function encodePrice(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodePriceParams(options).slice(2));
}
exports.encodePrice = encodePrice;
/**
 * Decodes the result of the price function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { decodePriceResult } from "thirdweb/extensions/farcaster";
 * const result = decodePriceResult("...");
 * ```
 */
function decodePriceResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodePriceResult = decodePriceResult;
/**
 * Calls the "price" function on the contract.
 * @param options - The options for the price function.
 * @returns The parsed result of the function call.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { price } from "thirdweb/extensions/farcaster";
 *
 * const result = await price({
 *  contract,
 *  extraStorage: ...,
 * });
 *
 * ```
 */
async function price(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.extraStorage],
    });
}
exports.price = price;
