// thirdweb/dist/cjs/extensions/erc2771/__generated__/IERC2771Context/read/isTrustedForwarder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTrustedForwarder = exports.decodeIsTrustedForwarderResult = exports.encodeIsTrustedForwarder = exports.encodeIsTrustedForwarderParams = exports.isIsTrustedForwarderSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x572b6c05";
const FN_INPUTS = [
    {
        type: "address",
        name: "forwarder",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `isTrustedForwarder` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `isTrustedForwarder` method is supported.
 * @extension ERC2771
 * @example
 * ```ts
 * import { isIsTrustedForwarderSupported } from "thirdweb/extensions/erc2771";
 *
 * const supported = await isIsTrustedForwarderSupported(contract);
 * ```
 */
async function isIsTrustedForwarderSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isIsTrustedForwarderSupported = isIsTrustedForwarderSupported;
/**
 * Encodes the parameters for the "isTrustedForwarder" function.
 * @param options - The options for the isTrustedForwarder function.
 * @returns The encoded ABI parameters.
 * @extension ERC2771
 * @example
 * ```ts
 * import { encodeIsTrustedForwarderParams } "thirdweb/extensions/erc2771";
 * const result = encodeIsTrustedForwarderParams({
 *  forwarder: ...,
 * });
 * ```
 */
function encodeIsTrustedForwarderParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.forwarder]);
}
exports.encodeIsTrustedForwarderParams = encodeIsTrustedForwarderParams;
/**
 * Encodes the "isTrustedForwarder" function into a Hex string with its parameters.
 * @param options - The options for the isTrustedForwarder function.
 * @returns The encoded hexadecimal string.
 * @extension ERC2771
 * @example
 * ```ts
 * import { encodeIsTrustedForwarder } "thirdweb/extensions/erc2771";
 * const result = encodeIsTrustedForwarder({
 *  forwarder: ...,
 * });
 * ```
 */
function encodeIsTrustedForwarder(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeIsTrustedForwarderParams(options).slice(2));
}
exports.encodeIsTrustedForwarder = encodeIsTrustedForwarder;
/**
 * Decodes the result of the isTrustedForwarder function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC2771
 * @example
 * ```ts
 * import { decodeIsTrustedForwarderResult } from "thirdweb/extensions/erc2771";
 * const result = decodeIsTrustedForwarderResult("...");
 * ```
 */
function decodeIsTrustedForwarderResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeIsTrustedForwarderResult = decodeIsTrustedForwarderResult;
/**
 * Calls the "isTrustedForwarder" function on the contract.
 * @param options - The options for the isTrustedForwarder function.
 * @returns The parsed result of the function call.
 * @extension ERC2771
 * @example
 * ```ts
 * import { isTrustedForwarder } from "thirdweb/extensions/erc2771";
 *
 * const result = await isTrustedForwarder({
 *  contract,
 *  forwarder: ...,
 * });
 *
 * ```
 */
async function isTrustedForwarder(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.forwarder],
    });
}
exports.isTrustedForwarder = isTrustedForwarder;
