// thirdweb/dist/cjs/exports/chains.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bscTestnet = exports.bsc = exports.zora = exports.zoraSepolia = exports.sepolia = exports.polygon = exports.polygonAmoy = exports.mumbai = exports.polygonMumbai = exports.optimism = exports.optimismSepolia = exports.mainnet = exports.ethereum = exports.base = exports.baseSepolia = exports.avalanche = exports.avalancheFuji = exports.arbitrum = exports.arbitrumSepolia = exports.arbitrumNova = exports.anvil = exports.getRpcUrlForChain = exports.getChainMetadata = exports.defineChain = void 0;
// define chain, chainMetadata
var utils_js_1 = require("../chains/utils.js");
Object.defineProperty(exports, "defineChain", { enumerable: true, get: function () { return utils_js_1.defineChain; } });
Object.defineProperty(exports, "getChainMetadata", { enumerable: true, get: function () { return utils_js_1.getChainMetadata; } });
Object.defineProperty(exports, "getRpcUrlForChain", { enumerable: true, get: function () { return utils_js_1.getRpcUrlForChain; } });
/**
 * PRE_DEFINED CHAINS
 */
var anvil_js_1 = require("../chains/chain-definitions/anvil.js");
Object.defineProperty(exports, "anvil", { enumerable: true, get: function () { return anvil_js_1.anvil; } });
var arbitrum_nova_js_1 = require("../chains/chain-definitions/arbitrum-nova.js");
Object.defineProperty(exports, "arbitrumNova", { enumerable: true, get: function () { return arbitrum_nova_js_1.arbitrumNova; } });
var arbitrum_sepolia_js_1 = require("../chains/chain-definitions/arbitrum-sepolia.js");
Object.defineProperty(exports, "arbitrumSepolia", { enumerable: true, get: function () { return arbitrum_sepolia_js_1.arbitrumSepolia; } });
var arbitrum_js_1 = require("../chains/chain-definitions/arbitrum.js");
Object.defineProperty(exports, "arbitrum", { enumerable: true, get: function () { return arbitrum_js_1.arbitrum; } });
var avalanche_fuji_js_1 = require("../chains/chain-definitions/avalanche-fuji.js");
Object.defineProperty(exports, "avalancheFuji", { enumerable: true, get: function () { return avalanche_fuji_js_1.avalancheFuji; } });
var avalanche_js_1 = require("../chains/chain-definitions/avalanche.js");
Object.defineProperty(exports, "avalanche", { enumerable: true, get: function () { return avalanche_js_1.avalanche; } });
var base_sepolia_js_1 = require("../chains/chain-definitions/base-sepolia.js");
Object.defineProperty(exports, "baseSepolia", { enumerable: true, get: function () { return base_sepolia_js_1.baseSepolia; } });
var base_js_1 = require("../chains/chain-definitions/base.js");
Object.defineProperty(exports, "base", { enumerable: true, get: function () { return base_js_1.base; } });
// mainnet = alias for ethereum
var ethereum_js_1 = require("../chains/chain-definitions/ethereum.js");
Object.defineProperty(exports, "ethereum", { enumerable: true, get: function () { return ethereum_js_1.ethereum; } });
Object.defineProperty(exports, "mainnet", { enumerable: true, get: function () { return ethereum_js_1.mainnet; } });
var optimism_sepolia_js_1 = require("../chains/chain-definitions/optimism-sepolia.js");
Object.defineProperty(exports, "optimismSepolia", { enumerable: true, get: function () { return optimism_sepolia_js_1.optimismSepolia; } });
var optimism_js_1 = require("../chains/chain-definitions/optimism.js");
Object.defineProperty(exports, "optimism", { enumerable: true, get: function () { return optimism_js_1.optimism; } });
// mumbai = alias for polygonMumbai
var polygon_mumbai_js_1 = require("../chains/chain-definitions/polygon-mumbai.js");
Object.defineProperty(exports, "polygonMumbai", { enumerable: true, get: function () { return polygon_mumbai_js_1.polygonMumbai; } });
Object.defineProperty(exports, "mumbai", { enumerable: true, get: function () { return polygon_mumbai_js_1.mumbai; } });
var polygon_amoy_js_1 = require("../chains/chain-definitions/polygon-amoy.js");
Object.defineProperty(exports, "polygonAmoy", { enumerable: true, get: function () { return polygon_amoy_js_1.polygonAmoy; } });
var polygon_js_1 = require("../chains/chain-definitions/polygon.js");
Object.defineProperty(exports, "polygon", { enumerable: true, get: function () { return polygon_js_1.polygon; } });
var sepolia_js_1 = require("../chains/chain-definitions/sepolia.js");
Object.defineProperty(exports, "sepolia", { enumerable: true, get: function () { return sepolia_js_1.sepolia; } });
var zora_sepolia_js_1 = require("../chains/chain-definitions/zora-sepolia.js");
Object.defineProperty(exports, "zoraSepolia", { enumerable: true, get: function () { return zora_sepolia_js_1.zoraSepolia; } });
var zora_js_1 = require("../chains/chain-definitions/zora.js");
Object.defineProperty(exports, "zora", { enumerable: true, get: function () { return zora_js_1.zora; } });
var bsc_js_1 = require("../chains/chain-definitions/bsc.js");
Object.defineProperty(exports, "bsc", { enumerable: true, get: function () { return bsc_js_1.bsc; } });
var bsc_testnet_js_1 = require("../chains/chain-definitions/bsc-testnet.js");
Object.defineProperty(exports, "bscTestnet", { enumerable: true, get: function () { return bsc_testnet_js_1.bscTestnet; } });
