// thirdweb/dist/cjs/extensions/airdrop/__generated__/Airdrop/read/tokenMerkleRoot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tokenMerkleRoot = exports.decodeTokenMerkleRootResult = exports.encodeTokenMerkleRoot = exports.encodeTokenMerkleRootParams = exports.isTokenMerkleRootSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x95f5c120";
const FN_INPUTS = [
    {
        type: "address",
        name: "tokenAddress",
    },
];
const FN_OUTPUTS = [
    {
        type: "bytes32",
    },
];
/**
 * Checks if the `tokenMerkleRoot` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `tokenMerkleRoot` method is supported.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { isTokenMerkleRootSupported } from "thirdweb/extensions/airdrop";
 *
 * const supported = await isTokenMerkleRootSupported(contract);
 * ```
 */
async function isTokenMerkleRootSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isTokenMerkleRootSupported = isTokenMerkleRootSupported;
/**
 * Encodes the parameters for the "tokenMerkleRoot" function.
 * @param options - The options for the tokenMerkleRoot function.
 * @returns The encoded ABI parameters.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { encodeTokenMerkleRootParams } "thirdweb/extensions/airdrop";
 * const result = encodeTokenMerkleRootParams({
 *  tokenAddress: ...,
 * });
 * ```
 */
function encodeTokenMerkleRootParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.tokenAddress]);
}
exports.encodeTokenMerkleRootParams = encodeTokenMerkleRootParams;
/**
 * Encodes the "tokenMerkleRoot" function into a Hex string with its parameters.
 * @param options - The options for the tokenMerkleRoot function.
 * @returns The encoded hexadecimal string.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { encodeTokenMerkleRoot } "thirdweb/extensions/airdrop";
 * const result = encodeTokenMerkleRoot({
 *  tokenAddress: ...,
 * });
 * ```
 */
function encodeTokenMerkleRoot(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeTokenMerkleRootParams(options).slice(2));
}
exports.encodeTokenMerkleRoot = encodeTokenMerkleRoot;
/**
 * Decodes the result of the tokenMerkleRoot function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { decodeTokenMerkleRootResult } from "thirdweb/extensions/airdrop";
 * const result = decodeTokenMerkleRootResult("...");
 * ```
 */
function decodeTokenMerkleRootResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeTokenMerkleRootResult = decodeTokenMerkleRootResult;
/**
 * Calls the "tokenMerkleRoot" function on the contract.
 * @param options - The options for the tokenMerkleRoot function.
 * @returns The parsed result of the function call.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { tokenMerkleRoot } from "thirdweb/extensions/airdrop";
 *
 * const result = await tokenMerkleRoot({
 *  contract,
 *  tokenAddress: ...,
 * });
 *
 * ```
 */
async function tokenMerkleRoot(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.tokenAddress],
    });
}
exports.tokenMerkleRoot = tokenMerkleRoot;
