// thirdweb/dist/cjs/react/web/ui/ConnectWallet/Modal/ConnectModalSkeleton.js
"use strict";
"use client";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectModalCompactLayout = exports.ConnectModalWideLayout = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const basic_js_1 = require("../../components/basic.js");
const CustomThemeProvider_js_1 = require("../../design-system/CustomThemeProvider.js");
const elements_js_1 = require("../../design-system/elements.js");
const constants_js_1 = require("../constants.js");
/**
 * @internal
 */
function ConnectModalWideLayout(props) {
    return ((0, jsx_runtime_1.jsxs)("div", { style: {
            height: "100%",
            display: "grid",
            gridTemplateColumns: "300px 1fr",
        }, children: [(0, jsx_runtime_1.jsxs)(LeftContainer, { children: [" ", props.left, " "] }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "column", scrollY: true, relative: true, children: props.right })] }));
}
exports.ConnectModalWideLayout = ConnectModalWideLayout;
/**
 * @internal
 */
function ConnectModalCompactLayout(props) {
    return ((0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "column", scrollY: true, relative: true, style: {
            maxHeight: constants_js_1.compactModalMaxHeight,
        }, children: props.children }));
}
exports.ConnectModalCompactLayout = ConnectModalCompactLayout;
const LeftContainer = /* @__PURE__ */ (0, elements_js_1.StyledDiv)(() => {
    const theme = (0, CustomThemeProvider_js_1.useCustomTheme)();
    return {
        display: "flex",
        flexDirection: "column",
        overflowY: "auto",
        ...basic_js_1.noScrollBar,
        position: "relative",
        borderRight: `1px solid ${theme.colors.separatorLine}`,
    };
});
