// thirdweb/dist/cjs/extensions/airdrop/__generated__/Airdrop/write/airdropERC1155.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.airdropERC1155 = exports.encodeAirdropERC1155 = exports.encodeAirdropERC1155Params = exports.isAirdropERC1155Supported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x2d89e38b";
const FN_INPUTS = [
    {
        name: "_tokenAddress",
        type: "address",
        internalType: "address",
    },
    {
        name: "_contents",
        type: "tuple[]",
        internalType: "struct Airdrop.AirdropContentERC1155[]",
        components: [
            {
                name: "recipient",
                type: "address",
                internalType: "address",
            },
            {
                name: "tokenId",
                type: "uint256",
                internalType: "uint256",
            },
            {
                name: "amount",
                type: "uint256",
                internalType: "uint256",
            },
        ],
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `airdropERC1155` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `airdropERC1155` method is supported.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { isAirdropERC1155Supported } from "thirdweb/extensions/airdrop";
 *
 * const supported = await isAirdropERC1155Supported(contract);
 * ```
 */
async function isAirdropERC1155Supported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isAirdropERC1155Supported = isAirdropERC1155Supported;
/**
 * Encodes the parameters for the "airdropERC1155" function.
 * @param options - The options for the airdropERC1155 function.
 * @returns The encoded ABI parameters.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { encodeAirdropERC1155Params } "thirdweb/extensions/airdrop";
 * const result = encodeAirdropERC1155Params({
 *  tokenAddress: ...,
 *  contents: ...,
 * });
 * ```
 */
function encodeAirdropERC1155Params(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [
        options.tokenAddress,
        options.contents,
    ]);
}
exports.encodeAirdropERC1155Params = encodeAirdropERC1155Params;
/**
 * Encodes the "airdropERC1155" function into a Hex string with its parameters.
 * @param options - The options for the airdropERC1155 function.
 * @returns The encoded hexadecimal string.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { encodeAirdropERC1155 } "thirdweb/extensions/airdrop";
 * const result = encodeAirdropERC1155({
 *  tokenAddress: ...,
 *  contents: ...,
 * });
 * ```
 */
function encodeAirdropERC1155(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeAirdropERC1155Params(options).slice(2));
}
exports.encodeAirdropERC1155 = encodeAirdropERC1155;
/**
 * Prepares a transaction to call the "airdropERC1155" function on the contract.
 * @param options - The options for the "airdropERC1155" function.
 * @returns A prepared transaction object.
 * @extension AIRDROP
 * @example
 * ```ts
 * import { airdropERC1155 } from "thirdweb/extensions/airdrop";
 *
 * const transaction = airdropERC1155({
 *  contract,
 *  tokenAddress: ...,
 *  contents: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function airdropERC1155(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.tokenAddress, resolvedOptions.contents];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.airdropERC1155 = airdropERC1155;
