// thirdweb/dist/cjs/extensions/thirdweb/__generated__/IThirdwebContract/read/contractVersion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contractVersion = exports.decodeContractVersionResult = exports.isContractVersionSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xa0a8e460";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint8",
    },
];
/**
 * Checks if the `contractVersion` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `contractVersion` method is supported.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { isContractVersionSupported } from "thirdweb/extensions/thirdweb";
 *
 * const supported = await isContractVersionSupported(contract);
 * ```
 */
async function isContractVersionSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isContractVersionSupported = isContractVersionSupported;
/**
 * Decodes the result of the contractVersion function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { decodeContractVersionResult } from "thirdweb/extensions/thirdweb";
 * const result = decodeContractVersionResult("...");
 * ```
 */
function decodeContractVersionResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeContractVersionResult = decodeContractVersionResult;
/**
 * Calls the "contractVersion" function on the contract.
 * @param options - The options for the contractVersion function.
 * @returns The parsed result of the function call.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { contractVersion } from "thirdweb/extensions/thirdweb";
 *
 * const result = await contractVersion({
 *  contract,
 * });
 *
 * ```
 */
async function contractVersion(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.contractVersion = contractVersion;
