// thirdweb/dist/esm/exports/react-native.js
export { ThirdwebProvider } from "../react/core/providers/thirdweb-provider.js";
// react/core
export { useSetActiveWallet, useActiveWalletChain, useConnect, useDisconnect, useActiveAccount, useActiveWallet, useConnectedWallets, useSwitchActiveWalletChain, useActiveWalletConnectionStatus, useSetActiveWalletConnectionStatus, useIsAutoConnecting, } from "../react/core/hooks/wallets/wallet-hooks.js";
// contract related
export { useReadContract } from "../react/core/hooks/contract/useReadContract.js";
export { useEstimateGas } from "../react/core/hooks/contract/useEstimateGas.js";
export { useWaitForReceipt } from "../react/core/hooks/contract/useWaitForReceipt.js";
export { useContractEvents } from "../react/core/hooks/contract/useContractEvents.js";
// rpc related
export { useBlockNumber, } from "../react/core/hooks/rpc/useBlockNumber.js";
// utils
export { createContractQuery } from "../react/core/utils/createQuery.js";
export { useInvalidateContractQuery } from "../react/core/hooks/others/useInvalidateQueries.js";
// pay
export { useBuyWithCryptoQuote, } from "../react/core/hooks/pay/useBuyWithCryptoQuote.js";
export { useBuyWithCryptoStatus } from "../react/core/hooks/pay/useBuyWithCryptoStatus.js";
export { useBuyWithCryptoHistory, } from "../react/core/hooks/pay/useBuyWithCryptoHistory.js";
export { useBuyWithFiatQuote, } from "../react/core/hooks/pay/useBuyWithFiatQuote.js";
export { useBuyWithFiatStatus } from "../react/core/hooks/pay/useBuyWithFiatStatus.js";
export { useBuyWithFiatHistory, } from "../react/core/hooks/pay/useBuyWithFiatHistory.js";
export { useBuyHistory, } from "../react/core/hooks/pay/useBuyHistory.js";
export { usePostOnRampQuote, } from "../react/core/hooks/pay/usePostOnrampQuote.js";
import { useSendTransactionCore } from "../react/core/hooks/contract/useSendTransaction.js";
/**
 * A hook to send a transaction.
 * @returns A mutation object to send a transaction.
 * @example
 * ```jsx
 * import { useSendTransaction } from "thirdweb/react-native";
 * const { mutate: sendTx, data: transactionResult } = useSendTransaction();
 *
 * // later
 * sendTx(tx);
 * ```
 * @transaction
 */
export const useSendTransaction = useSendTransactionCore;
export { useSendBatchTransaction } from "../react/core/hooks/contract/useSendBatchTransaction.js";
export { useSendAndConfirmTransaction } from "../react/core/hooks/contract/useSendAndConfirmTransaction.js";
export { useEstimateGasCost } from "../react/core/hooks/contract/useEstimateGasCost.js";
