// thirdweb/dist/esm/extensions/thirdweb/__generated__/IThirdwebContract/read/contractURI.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xe8a3d485";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "string",
    },
];
/**
 * Checks if the `contractURI` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `contractURI` method is supported.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { isContractURISupported } from "thirdweb/extensions/thirdweb";
 *
 * const supported = await isContractURISupported(contract);
 * ```
 */
export async function isContractURISupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the contractURI function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { decodeContractURIResult } from "thirdweb/extensions/thirdweb";
 * const result = decodeContractURIResult("...");
 * ```
 */
export function decodeContractURIResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "contractURI" function on the contract.
 * @param options - The options for the contractURI function.
 * @returns The parsed result of the function call.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { contractURI } from "thirdweb/extensions/thirdweb";
 *
 * const result = await contractURI({
 *  contract,
 * });
 *
 * ```
 */
export async function contractURI(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
