// thirdweb/dist/cjs/extensions/erc4337/__generated__/IAccountFactory/read/getAccountsOfSigner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAccountsOfSigner = exports.decodeGetAccountsOfSignerResult = exports.encodeGetAccountsOfSigner = exports.encodeGetAccountsOfSignerParams = exports.isGetAccountsOfSignerSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x0e6254fd";
const FN_INPUTS = [
    {
        type: "address",
        name: "signer",
    },
];
const FN_OUTPUTS = [
    {
        type: "address[]",
        name: "accounts",
    },
];
/**
 * Checks if the `getAccountsOfSigner` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getAccountsOfSigner` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isGetAccountsOfSignerSupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = await isGetAccountsOfSignerSupported(contract);
 * ```
 */
async function isGetAccountsOfSignerSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetAccountsOfSignerSupported = isGetAccountsOfSignerSupported;
/**
 * Encodes the parameters for the "getAccountsOfSigner" function.
 * @param options - The options for the getAccountsOfSigner function.
 * @returns The encoded ABI parameters.
 * @extension ERC4337
 * @example
 * ```ts
 * import { encodeGetAccountsOfSignerParams } "thirdweb/extensions/erc4337";
 * const result = encodeGetAccountsOfSignerParams({
 *  signer: ...,
 * });
 * ```
 */
function encodeGetAccountsOfSignerParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.signer]);
}
exports.encodeGetAccountsOfSignerParams = encodeGetAccountsOfSignerParams;
/**
 * Encodes the "getAccountsOfSigner" function into a Hex string with its parameters.
 * @param options - The options for the getAccountsOfSigner function.
 * @returns The encoded hexadecimal string.
 * @extension ERC4337
 * @example
 * ```ts
 * import { encodeGetAccountsOfSigner } "thirdweb/extensions/erc4337";
 * const result = encodeGetAccountsOfSigner({
 *  signer: ...,
 * });
 * ```
 */
function encodeGetAccountsOfSigner(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetAccountsOfSignerParams(options).slice(2));
}
exports.encodeGetAccountsOfSigner = encodeGetAccountsOfSigner;
/**
 * Decodes the result of the getAccountsOfSigner function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC4337
 * @example
 * ```ts
 * import { decodeGetAccountsOfSignerResult } from "thirdweb/extensions/erc4337";
 * const result = decodeGetAccountsOfSignerResult("...");
 * ```
 */
function decodeGetAccountsOfSignerResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetAccountsOfSignerResult = decodeGetAccountsOfSignerResult;
/**
 * Calls the "getAccountsOfSigner" function on the contract.
 * @param options - The options for the getAccountsOfSigner function.
 * @returns The parsed result of the function call.
 * @extension ERC4337
 * @example
 * ```ts
 * import { getAccountsOfSigner } from "thirdweb/extensions/erc4337";
 *
 * const result = await getAccountsOfSigner({
 *  contract,
 *  signer: ...,
 * });
 *
 * ```
 */
async function getAccountsOfSigner(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.signer],
    });
}
exports.getAccountsOfSigner = getAccountsOfSigner;
