// thirdweb/dist/esm/adapters/wallet-adapter.js
import { createWalletEmitter } from "../wallets/wallet-emitter.js";
export function createWalletAdapter(options) {
    const emitter = createWalletEmitter();
    let _chain = options.chain;
    return {
        id: "adapter",
        subscribe: emitter.subscribe,
        connect: async () => {
            emitter.emit("onConnect", options);
            return options.adaptedAccount;
        },
        autoConnect: async () => {
            emitter.emit("onConnect", options);
            return options.adaptedAccount;
        },
        disconnect: async () => {
            await options.onDisconnect();
            emitter.emit("disconnect", undefined);
        },
        getAccount() {
            return options.adaptedAccount;
        },
        getChain() {
            return _chain;
        },
        getConfig() {
            return options;
        },
        switchChain: async (chain) => {
            await options.switchChain(chain);
            _chain = chain;
            emitter.emit("chainChanged", chain);
        },
    };
}
