// thirdweb/dist/cjs/adapters/wallet-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWalletAdapter = void 0;
const wallet_emitter_js_1 = require("../wallets/wallet-emitter.js");
function createWalletAdapter(options) {
    const emitter = (0, wallet_emitter_js_1.createWalletEmitter)();
    let _chain = options.chain;
    return {
        id: "adapter",
        subscribe: emitter.subscribe,
        connect: async () => {
            emitter.emit("onConnect", options);
            return options.adaptedAccount;
        },
        autoConnect: async () => {
            emitter.emit("onConnect", options);
            return options.adaptedAccount;
        },
        disconnect: async () => {
            await options.onDisconnect();
            emitter.emit("disconnect", undefined);
        },
        getAccount() {
            return options.adaptedAccount;
        },
        getChain() {
            return _chain;
        },
        getConfig() {
            return options;
        },
        switchChain: async (chain) => {
            await options.switchChain(chain);
            _chain = chain;
            emitter.emit("chainChanged", chain);
        },
    };
}
exports.createWalletAdapter = createWalletAdapter;
