// thirdweb/dist/cjs/extensions/erc20/__generated__/IVotes/read/getPastVotes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPastVotes = exports.decodeGetPastVotesResult = exports.encodeGetPastVotes = exports.encodeGetPastVotesParams = exports.isGetPastVotesSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x3a46b1a8";
const FN_INPUTS = [
    {
        type: "address",
        name: "account",
    },
    {
        type: "uint256",
        name: "blockNumber",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getPastVotes` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getPastVotes` method is supported.
 * @extension ERC20
 * @example
 * ```ts
 * import { isGetPastVotesSupported } from "thirdweb/extensions/erc20";
 *
 * const supported = await isGetPastVotesSupported(contract);
 * ```
 */
async function isGetPastVotesSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetPastVotesSupported = isGetPastVotesSupported;
/**
 * Encodes the parameters for the "getPastVotes" function.
 * @param options - The options for the getPastVotes function.
 * @returns The encoded ABI parameters.
 * @extension ERC20
 * @example
 * ```ts
 * import { encodeGetPastVotesParams } "thirdweb/extensions/erc20";
 * const result = encodeGetPastVotesParams({
 *  account: ...,
 *  blockNumber: ...,
 * });
 * ```
 */
function encodeGetPastVotesParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.account, options.blockNumber]);
}
exports.encodeGetPastVotesParams = encodeGetPastVotesParams;
/**
 * Encodes the "getPastVotes" function into a Hex string with its parameters.
 * @param options - The options for the getPastVotes function.
 * @returns The encoded hexadecimal string.
 * @extension ERC20
 * @example
 * ```ts
 * import { encodeGetPastVotes } "thirdweb/extensions/erc20";
 * const result = encodeGetPastVotes({
 *  account: ...,
 *  blockNumber: ...,
 * });
 * ```
 */
function encodeGetPastVotes(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetPastVotesParams(options).slice(2));
}
exports.encodeGetPastVotes = encodeGetPastVotes;
/**
 * Decodes the result of the getPastVotes function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC20
 * @example
 * ```ts
 * import { decodeGetPastVotesResult } from "thirdweb/extensions/erc20";
 * const result = decodeGetPastVotesResult("...");
 * ```
 */
function decodeGetPastVotesResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetPastVotesResult = decodeGetPastVotesResult;
/**
 * Calls the "getPastVotes" function on the contract.
 * @param options - The options for the getPastVotes function.
 * @returns The parsed result of the function call.
 * @extension ERC20
 * @example
 * ```ts
 * import { getPastVotes } from "thirdweb/extensions/erc20";
 *
 * const result = await getPastVotes({
 *  contract,
 *  account: ...,
 *  blockNumber: ...,
 * });
 *
 * ```
 */
async function getPastVotes(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.account, options.blockNumber],
    });
}
exports.getPastVotes = getPastVotes;
