// thirdweb/dist/cjs/extensions/erc20/__generated__/IVotes/read/getPastTotalSupply.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPastTotalSupply = exports.decodeGetPastTotalSupplyResult = exports.encodeGetPastTotalSupply = exports.encodeGetPastTotalSupplyParams = exports.isGetPastTotalSupplySupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x8e539e8c";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "blockNumber",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getPastTotalSupply` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getPastTotalSupply` method is supported.
 * @extension ERC20
 * @example
 * ```ts
 * import { isGetPastTotalSupplySupported } from "thirdweb/extensions/erc20";
 *
 * const supported = await isGetPastTotalSupplySupported(contract);
 * ```
 */
async function isGetPastTotalSupplySupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetPastTotalSupplySupported = isGetPastTotalSupplySupported;
/**
 * Encodes the parameters for the "getPastTotalSupply" function.
 * @param options - The options for the getPastTotalSupply function.
 * @returns The encoded ABI parameters.
 * @extension ERC20
 * @example
 * ```ts
 * import { encodeGetPastTotalSupplyParams } "thirdweb/extensions/erc20";
 * const result = encodeGetPastTotalSupplyParams({
 *  blockNumber: ...,
 * });
 * ```
 */
function encodeGetPastTotalSupplyParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.blockNumber]);
}
exports.encodeGetPastTotalSupplyParams = encodeGetPastTotalSupplyParams;
/**
 * Encodes the "getPastTotalSupply" function into a Hex string with its parameters.
 * @param options - The options for the getPastTotalSupply function.
 * @returns The encoded hexadecimal string.
 * @extension ERC20
 * @example
 * ```ts
 * import { encodeGetPastTotalSupply } "thirdweb/extensions/erc20";
 * const result = encodeGetPastTotalSupply({
 *  blockNumber: ...,
 * });
 * ```
 */
function encodeGetPastTotalSupply(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetPastTotalSupplyParams(options).slice(2));
}
exports.encodeGetPastTotalSupply = encodeGetPastTotalSupply;
/**
 * Decodes the result of the getPastTotalSupply function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC20
 * @example
 * ```ts
 * import { decodeGetPastTotalSupplyResult } from "thirdweb/extensions/erc20";
 * const result = decodeGetPastTotalSupplyResult("...");
 * ```
 */
function decodeGetPastTotalSupplyResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetPastTotalSupplyResult = decodeGetPastTotalSupplyResult;
/**
 * Calls the "getPastTotalSupply" function on the contract.
 * @param options - The options for the getPastTotalSupply function.
 * @returns The parsed result of the function call.
 * @extension ERC20
 * @example
 * ```ts
 * import { getPastTotalSupply } from "thirdweb/extensions/erc20";
 *
 * const result = await getPastTotalSupply({
 *  contract,
 *  blockNumber: ...,
 * });
 *
 * ```
 */
async function getPastTotalSupply(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.blockNumber],
    });
}
exports.getPastTotalSupply = getPastTotalSupply;
