// thirdweb/dist/cjs/wallets/in-app/implementations/interfaces/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthProvider = exports.RecoveryShareManagement = void 0;
// TODO: remove usage of enums, instead use object with as const
var RecoveryShareManagement;
(function (RecoveryShareManagement) {
    RecoveryShareManagement["USER_MANAGED"] = "USER_MANAGED";
    RecoveryShareManagement["CLOUD_MANAGED"] = "AWS_MANAGED";
})(RecoveryShareManagement || (exports.RecoveryShareManagement = RecoveryShareManagement = {}));
// TODO: remove usage of enums, instead use object with as const
var AuthProvider;
(function (AuthProvider) {
    AuthProvider["COGNITO"] = "Cognito";
    AuthProvider["GOOGLE"] = "Google";
    AuthProvider["EMAIL_OTP"] = "EmailOtp";
    AuthProvider["CUSTOM_JWT"] = "CustomJWT";
    AuthProvider["CUSTOM_AUTH_ENDPOINT"] = "CustomAuthEndpoint";
    AuthProvider["FACEBOOK"] = "Facebook";
    AuthProvider["APPLE"] = "Apple";
    AuthProvider["PASSKEY"] = "Passkey";
})(AuthProvider || (exports.AuthProvider = AuthProvider = {}));
