// thirdweb/dist/cjs/extensions/erc721/write/sigMint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateMintSignature = exports.mintWithSignature = void 0;
const addresses_js_1 = require("../../../constants/addresses.js");
const bigint_js_1 = require("../../../utils/bigint.js");
const date_js_1 = require("../../../utils/date.js");
const random_js_1 = require("../../../utils/random.js");
const mintWithSignature_js_1 = require("../__generated__/ISignatureMintERC721/write/mintWithSignature.js");
/**
 * Mints a new ERC721 token with the given minter signature
 * @param options - The transaction options.
 * @example
 * ```ts
 * import { mintWithSignature, generateMintSignature } from "thirdweb/extensions/erc721";
 *
 * const { payload, signature } = await generateMintSignature(...)
 *
 * const transaction = mintWithSignature({
 *   contract,
 *   payload,
 *   signature,
 * });
 * await sendTransaction({ transaction, account });
 * ```
 * @extension ERC721
 * @returns A promise that resolves to the transaction result.
 */
function mintWithSignature(options) {
    const value = (0, addresses_js_1.isNativeTokenAddress)(options.payload.currency)
        ? options.payload.price
        : 0n;
    return (0, mintWithSignature_js_1.mintWithSignature)({
        ...options,
        overrides: {
            value,
        },
    });
}
exports.mintWithSignature = mintWithSignature;
/**
 * Generates the payload and signature for minting an ERC721 token.
 * @param options - The options for the minting process.
 * @example
 * ```ts
 * import { mintWithSignature, generateMintSignature } from "thirdweb/extensions/erc721";
 *
 * const { payload, signature } = await generateMintSignature({
 *   account,
 *   contract,
 *   mintRequest: {
 *     to: "0x...",
 *     metadata: {
 *       name: "My NFT",
 *       description: "This is my NFT",
 *       image: "https://example.com/image.png",
 *     },
 *   },
 * });
 *
 * const transaction = mintWithSignature({
 *   contract,
 *   payload,
 *   signature,
 * });
 * await sendTransaction({ transaction, account });
 * ```
 * @extension ERC721
 * @returns A promise that resolves to the payload and signature.
 */
async function generateMintSignature(options) {
    const { mintRequest, account, contract } = options;
    const currency = mintRequest.currency || addresses_js_1.NATIVE_TOKEN_ADDRESS;
    const [price, uri, uid] = await Promise.all([
        // price per token in wei
        (async () => {
            // if priceInWei is provided, use it
            if ("priceInWei" in mintRequest && mintRequest.priceInWei) {
                return mintRequest.priceInWei;
            }
            // if price is provided, convert it to wei
            if ("price" in mintRequest && mintRequest.price) {
                const { convertErc20Amount } = await Promise.resolve().then(() => require("../../../utils/extensions/convert-erc20-amount.js"));
                return await convertErc20Amount({
                    amount: mintRequest.price,
                    client: contract.client,
                    chain: contract.chain,
                    erc20Address: currency,
                });
            }
            // if neither price nor priceInWei is provided, default to 0
            return 0n;
        })(),
        // uri
        (async () => {
            if ("metadata" in mintRequest) {
                if (typeof mintRequest.metadata === "object") {
                    // async import the upload function because it is not always required
                    const { upload } = await Promise.resolve().then(() => require("../../../storage/upload.js"));
                    return await upload({
                        client: options.contract.client,
                        files: [mintRequest.metadata],
                    });
                }
                return mintRequest.metadata;
            }
            return "";
        })(),
        // uid computation
        mintRequest.uid || (await (0, random_js_1.randomBytes)()),
    ]);
    const startTime = mintRequest.validityStartTimestamp || new Date(0);
    const endTime = mintRequest.validityEndTimestamp || (0, date_js_1.tenYearsFromNow)();
    const payload = {
        uri,
        currency,
        uid,
        price,
        to: mintRequest.to,
        royaltyRecipient: mintRequest.royaltyRecipient || account.address,
        royaltyBps: (0, bigint_js_1.toBigInt)(mintRequest.royaltyBps || 0),
        primarySaleRecipient: mintRequest.primarySaleRecipient || account.address,
        validityStartTimestamp: (0, date_js_1.dateToSeconds)(startTime),
        validityEndTimestamp: (0, date_js_1.dateToSeconds)(endTime),
    };
    const signature = await account.signTypedData({
        domain: {
            name: "TokenERC721",
            version: "1",
            chainId: contract.chain.id,
            verifyingContract: contract.address,
        },
        types: { MintRequest: MintRequest721 },
        primaryType: "MintRequest",
        message: payload,
    });
    return { payload, signature };
}
exports.generateMintSignature = generateMintSignature;
const MintRequest721 = [
    { name: "to", type: "address" },
    { name: "royaltyRecipient", type: "address" },
    { name: "royaltyBps", type: "uint256" },
    { name: "primarySaleRecipient", type: "address" },
    { name: "uri", type: "string" },
    { name: "price", type: "uint256" },
    { name: "currency", type: "address" },
    { name: "validityStartTimestamp", type: "uint128" },
    { name: "validityEndTimestamp", type: "uint128" },
    { name: "uid", type: "bytes32" },
];
