// thirdweb/dist/cjs/extensions/erc20/__generated__/IDropERC20/read/getActiveClaimConditionId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getActiveClaimConditionId = exports.decodeGetActiveClaimConditionIdResult = exports.isGetActiveClaimConditionIdSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xc68907de";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getActiveClaimConditionId` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getActiveClaimConditionId` method is supported.
 * @extension ERC20
 * @example
 * ```ts
 * import { isGetActiveClaimConditionIdSupported } from "thirdweb/extensions/erc20";
 *
 * const supported = await isGetActiveClaimConditionIdSupported(contract);
 * ```
 */
async function isGetActiveClaimConditionIdSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetActiveClaimConditionIdSupported = isGetActiveClaimConditionIdSupported;
/**
 * Decodes the result of the getActiveClaimConditionId function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC20
 * @example
 * ```ts
 * import { decodeGetActiveClaimConditionIdResult } from "thirdweb/extensions/erc20";
 * const result = decodeGetActiveClaimConditionIdResult("...");
 * ```
 */
function decodeGetActiveClaimConditionIdResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetActiveClaimConditionIdResult = decodeGetActiveClaimConditionIdResult;
/**
 * Calls the "getActiveClaimConditionId" function on the contract.
 * @param options - The options for the getActiveClaimConditionId function.
 * @returns The parsed result of the function call.
 * @extension ERC20
 * @example
 * ```ts
 * import { getActiveClaimConditionId } from "thirdweb/extensions/erc20";
 *
 * const result = await getActiveClaimConditionId({
 *  contract,
 * });
 *
 * ```
 */
async function getActiveClaimConditionId(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.getActiveClaimConditionId = getActiveClaimConditionId;
