// thirdweb/dist/cjs/extensions/erc20/__generated__/IVotes/write/delegateBySig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.delegateBySig = exports.encodeDelegateBySig = exports.encodeDelegateBySigParams = exports.isDelegateBySigSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xc3cda520";
const FN_INPUTS = [
    {
        type: "address",
        name: "delegatee",
    },
    {
        type: "uint256",
        name: "nonce",
    },
    {
        type: "uint256",
        name: "expiry",
    },
    {
        type: "uint8",
        name: "v",
    },
    {
        type: "bytes32",
        name: "r",
    },
    {
        type: "bytes32",
        name: "s",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `delegateBySig` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `delegateBySig` method is supported.
 * @extension ERC20
 * @example
 * ```ts
 * import { isDelegateBySigSupported } from "thirdweb/extensions/erc20";
 *
 * const supported = await isDelegateBySigSupported(contract);
 * ```
 */
async function isDelegateBySigSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isDelegateBySigSupported = isDelegateBySigSupported;
/**
 * Encodes the parameters for the "delegateBySig" function.
 * @param options - The options for the delegateBySig function.
 * @returns The encoded ABI parameters.
 * @extension ERC20
 * @example
 * ```ts
 * import { encodeDelegateBySigParams } "thirdweb/extensions/erc20";
 * const result = encodeDelegateBySigParams({
 *  delegatee: ...,
 *  nonce: ...,
 *  expiry: ...,
 *  v: ...,
 *  r: ...,
 *  s: ...,
 * });
 * ```
 */
function encodeDelegateBySigParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [
        options.delegatee,
        options.nonce,
        options.expiry,
        options.v,
        options.r,
        options.s,
    ]);
}
exports.encodeDelegateBySigParams = encodeDelegateBySigParams;
/**
 * Encodes the "delegateBySig" function into a Hex string with its parameters.
 * @param options - The options for the delegateBySig function.
 * @returns The encoded hexadecimal string.
 * @extension ERC20
 * @example
 * ```ts
 * import { encodeDelegateBySig } "thirdweb/extensions/erc20";
 * const result = encodeDelegateBySig({
 *  delegatee: ...,
 *  nonce: ...,
 *  expiry: ...,
 *  v: ...,
 *  r: ...,
 *  s: ...,
 * });
 * ```
 */
function encodeDelegateBySig(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeDelegateBySigParams(options).slice(2));
}
exports.encodeDelegateBySig = encodeDelegateBySig;
/**
 * Prepares a transaction to call the "delegateBySig" function on the contract.
 * @param options - The options for the "delegateBySig" function.
 * @returns A prepared transaction object.
 * @extension ERC20
 * @example
 * ```ts
 * import { delegateBySig } from "thirdweb/extensions/erc20";
 *
 * const transaction = delegateBySig({
 *  contract,
 *  delegatee: ...,
 *  nonce: ...,
 *  expiry: ...,
 *  v: ...,
 *  r: ...,
 *  s: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function delegateBySig(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [
                resolvedOptions.delegatee,
                resolvedOptions.nonce,
                resolvedOptions.expiry,
                resolvedOptions.v,
                resolvedOptions.r,
                resolvedOptions.s,
            ];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.delegateBySig = delegateBySig;
