// thirdweb/dist/esm/extensions/erc4337/account/removeSessionKey.js
import { setPermissionsForSigner } from "../__generated__/IAccountPermissions/write/setPermissionsForSigner.js";
import { signPermissionRequest, toContractPermissions } from "./common.js";
/**
 * Removes session key permissions for a specified address.
 * @param options - The options for the removeSessionKey function.
 * @returns The transaction object to be sent.
 * @example
 * ```ts
 * import { removeSessionKey } from 'thirdweb/extensions/erc4337';
 *
 * const transaction = removeSessionKey({
 * contract,
 * account,
 * sessionKeyAddress
 * });
 * await sendTransaction({ transaction, account });
 * ```
 * @extension ERC4337
 */
export function removeSessionKey(options) {
    const { contract, account, sessionKeyAddress } = options;
    return setPermissionsForSigner({
        contract,
        async asyncParams() {
            const { req, signature } = await signPermissionRequest({
                account,
                contract,
                req: await toContractPermissions({
                    target: sessionKeyAddress,
                    permissions: {
                        approvedTargets: [],
                        nativeTokenLimitPerTransaction: 0,
                        permissionStartTimestamp: new Date(0),
                        permissionEndTimestamp: new Date(0),
                    },
                }),
            });
            return { signature, req };
        },
    });
}
