// thirdweb/dist/cjs/extensions/permissions/__generated__/IPermissionsEnumerable/read/getRoleMember.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRoleMember = exports.decodeGetRoleMemberResult = exports.encodeGetRoleMember = exports.encodeGetRoleMemberParams = exports.isGetRoleMemberSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x9010d07c";
const FN_INPUTS = [
    {
        type: "bytes32",
        name: "role",
    },
    {
        type: "uint256",
        name: "index",
    },
];
const FN_OUTPUTS = [
    {
        type: "address",
    },
];
/**
 * Checks if the `getRoleMember` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getRoleMember` method is supported.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { isGetRoleMemberSupported } from "thirdweb/extensions/permissions";
 *
 * const supported = await isGetRoleMemberSupported(contract);
 * ```
 */
async function isGetRoleMemberSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetRoleMemberSupported = isGetRoleMemberSupported;
/**
 * Encodes the parameters for the "getRoleMember" function.
 * @param options - The options for the getRoleMember function.
 * @returns The encoded ABI parameters.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { encodeGetRoleMemberParams } "thirdweb/extensions/permissions";
 * const result = encodeGetRoleMemberParams({
 *  role: ...,
 *  index: ...,
 * });
 * ```
 */
function encodeGetRoleMemberParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.role, options.index]);
}
exports.encodeGetRoleMemberParams = encodeGetRoleMemberParams;
/**
 * Encodes the "getRoleMember" function into a Hex string with its parameters.
 * @param options - The options for the getRoleMember function.
 * @returns The encoded hexadecimal string.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { encodeGetRoleMember } "thirdweb/extensions/permissions";
 * const result = encodeGetRoleMember({
 *  role: ...,
 *  index: ...,
 * });
 * ```
 */
function encodeGetRoleMember(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetRoleMemberParams(options).slice(2));
}
exports.encodeGetRoleMember = encodeGetRoleMember;
/**
 * Decodes the result of the getRoleMember function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { decodeGetRoleMemberResult } from "thirdweb/extensions/permissions";
 * const result = decodeGetRoleMemberResult("...");
 * ```
 */
function decodeGetRoleMemberResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetRoleMemberResult = decodeGetRoleMemberResult;
/**
 * Calls the "getRoleMember" function on the contract.
 * @param options - The options for the getRoleMember function.
 * @returns The parsed result of the function call.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { getRoleMember } from "thirdweb/extensions/permissions";
 *
 * const result = await getRoleMember({
 *  contract,
 *  role: ...,
 *  index: ...,
 * });
 *
 * ```
 */
async function getRoleMember(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.role, options.index],
    });
}
exports.getRoleMember = getRoleMember;
