// thirdweb/dist/cjs/extensions/permissions/__generated__/IPermissionsEnumerable/read/hasRole.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasRole = exports.decodeHasRoleResult = exports.encodeHasRole = exports.encodeHasRoleParams = exports.isHasRoleSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x91d14854";
const FN_INPUTS = [
    {
        type: "bytes32",
        name: "role",
    },
    {
        type: "address",
        name: "account",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `hasRole` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `hasRole` method is supported.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { isHasRoleSupported } from "thirdweb/extensions/permissions";
 *
 * const supported = await isHasRoleSupported(contract);
 * ```
 */
async function isHasRoleSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isHasRoleSupported = isHasRoleSupported;
/**
 * Encodes the parameters for the "hasRole" function.
 * @param options - The options for the hasRole function.
 * @returns The encoded ABI parameters.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { encodeHasRoleParams } "thirdweb/extensions/permissions";
 * const result = encodeHasRoleParams({
 *  role: ...,
 *  account: ...,
 * });
 * ```
 */
function encodeHasRoleParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.role, options.account]);
}
exports.encodeHasRoleParams = encodeHasRoleParams;
/**
 * Encodes the "hasRole" function into a Hex string with its parameters.
 * @param options - The options for the hasRole function.
 * @returns The encoded hexadecimal string.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { encodeHasRole } "thirdweb/extensions/permissions";
 * const result = encodeHasRole({
 *  role: ...,
 *  account: ...,
 * });
 * ```
 */
function encodeHasRole(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeHasRoleParams(options).slice(2));
}
exports.encodeHasRole = encodeHasRole;
/**
 * Decodes the result of the hasRole function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { decodeHasRoleResult } from "thirdweb/extensions/permissions";
 * const result = decodeHasRoleResult("...");
 * ```
 */
function decodeHasRoleResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeHasRoleResult = decodeHasRoleResult;
/**
 * Calls the "hasRole" function on the contract.
 * @param options - The options for the hasRole function.
 * @returns The parsed result of the function call.
 * @extension PERMISSIONS
 * @example
 * ```ts
 * import { hasRole } from "thirdweb/extensions/permissions";
 *
 * const result = await hasRole({
 *  contract,
 *  role: ...,
 *  account: ...,
 * });
 *
 * ```
 */
async function hasRole(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.role, options.account],
    });
}
exports.hasRole = hasRole;
