// thirdweb/dist/cjs/extensions/erc721/__generated__/INFTStake/write/withdrawRewardTokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withdrawRewardTokens = exports.encodeWithdrawRewardTokens = exports.encodeWithdrawRewardTokensParams = exports.isWithdrawRewardTokensSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xcb43b2dd";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_amount",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `withdrawRewardTokens` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `withdrawRewardTokens` method is supported.
 * @extension ERC721
 * @example
 * ```ts
 * import { isWithdrawRewardTokensSupported } from "thirdweb/extensions/erc721";
 *
 * const supported = await isWithdrawRewardTokensSupported(contract);
 * ```
 */
async function isWithdrawRewardTokensSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isWithdrawRewardTokensSupported = isWithdrawRewardTokensSupported;
/**
 * Encodes the parameters for the "withdrawRewardTokens" function.
 * @param options - The options for the withdrawRewardTokens function.
 * @returns The encoded ABI parameters.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeWithdrawRewardTokensParams } "thirdweb/extensions/erc721";
 * const result = encodeWithdrawRewardTokensParams({
 *  amount: ...,
 * });
 * ```
 */
function encodeWithdrawRewardTokensParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.amount]);
}
exports.encodeWithdrawRewardTokensParams = encodeWithdrawRewardTokensParams;
/**
 * Encodes the "withdrawRewardTokens" function into a Hex string with its parameters.
 * @param options - The options for the withdrawRewardTokens function.
 * @returns The encoded hexadecimal string.
 * @extension ERC721
 * @example
 * ```ts
 * import { encodeWithdrawRewardTokens } "thirdweb/extensions/erc721";
 * const result = encodeWithdrawRewardTokens({
 *  amount: ...,
 * });
 * ```
 */
function encodeWithdrawRewardTokens(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeWithdrawRewardTokensParams(options).slice(2));
}
exports.encodeWithdrawRewardTokens = encodeWithdrawRewardTokens;
/**
 * Prepares a transaction to call the "withdrawRewardTokens" function on the contract.
 * @param options - The options for the "withdrawRewardTokens" function.
 * @returns A prepared transaction object.
 * @extension ERC721
 * @example
 * ```ts
 * import { withdrawRewardTokens } from "thirdweb/extensions/erc721";
 *
 * const transaction = withdrawRewardTokens({
 *  contract,
 *  amount: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function withdrawRewardTokens(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.amount];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.withdrawRewardTokens = withdrawRewardTokens;
