// thirdweb/dist/esm/extensions/erc721/__generated__/IERC721A/read/symbol.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x95d89b41";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "string",
    },
];
/**
 * Checks if the `symbol` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `symbol` method is supported.
 * @extension ERC721
 * @example
 * ```ts
 * import { isSymbolSupported } from "thirdweb/extensions/erc721";
 *
 * const supported = await isSymbolSupported(contract);
 * ```
 */
export async function isSymbolSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the symbol function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC721
 * @example
 * ```ts
 * import { decodeSymbolResult } from "thirdweb/extensions/erc721";
 * const result = decodeSymbolResult("...");
 * ```
 */
export function decodeSymbolResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "symbol" function on the contract.
 * @param options - The options for the symbol function.
 * @returns The parsed result of the function call.
 * @extension ERC721
 * @example
 * ```ts
 * import { symbol } from "thirdweb/extensions/erc721";
 *
 * const result = await symbol({
 *  contract,
 * });
 *
 * ```
 */
export async function symbol(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
