// thirdweb/dist/cjs/extensions/farcaster/__generated__/IIdGateway/read/idRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.idRegistry = exports.decodeIdRegistryResult = exports.isIdRegistrySupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x0aa13b8c";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "address",
    },
];
/**
 * Checks if the `idRegistry` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `idRegistry` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isIdRegistrySupported } from "thirdweb/extensions/farcaster";
 *
 * const supported = await isIdRegistrySupported(contract);
 * ```
 */
async function isIdRegistrySupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isIdRegistrySupported = isIdRegistrySupported;
/**
 * Decodes the result of the idRegistry function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { decodeIdRegistryResult } from "thirdweb/extensions/farcaster";
 * const result = decodeIdRegistryResult("...");
 * ```
 */
function decodeIdRegistryResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeIdRegistryResult = decodeIdRegistryResult;
/**
 * Calls the "idRegistry" function on the contract.
 * @param options - The options for the idRegistry function.
 * @returns The parsed result of the function call.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { idRegistry } from "thirdweb/extensions/farcaster";
 *
 * const result = await idRegistry({
 *  contract,
 * });
 *
 * ```
 */
async function idRegistry(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.idRegistry = idRegistry;
