// thirdweb/dist/cjs/extensions/erc721/__generated__/ISharedMetadata/read/sharedMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sharedMetadata = exports.decodeSharedMetadataResult = exports.isSharedMetadataSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xb280f703";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "string",
        name: "name",
    },
    {
        type: "string",
        name: "description",
    },
    {
        type: "string",
        name: "imageURI",
    },
    {
        type: "string",
        name: "animationURI",
    },
];
/**
 * Checks if the `sharedMetadata` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `sharedMetadata` method is supported.
 * @extension ERC721
 * @example
 * ```ts
 * import { isSharedMetadataSupported } from "thirdweb/extensions/erc721";
 *
 * const supported = await isSharedMetadataSupported(contract);
 * ```
 */
async function isSharedMetadataSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isSharedMetadataSupported = isSharedMetadataSupported;
/**
 * Decodes the result of the sharedMetadata function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC721
 * @example
 * ```ts
 * import { decodeSharedMetadataResult } from "thirdweb/extensions/erc721";
 * const result = decodeSharedMetadataResult("...");
 * ```
 */
function decodeSharedMetadataResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result);
}
exports.decodeSharedMetadataResult = decodeSharedMetadataResult;
/**
 * Calls the "sharedMetadata" function on the contract.
 * @param options - The options for the sharedMetadata function.
 * @returns The parsed result of the function call.
 * @extension ERC721
 * @example
 * ```ts
 * import { sharedMetadata } from "thirdweb/extensions/erc721";
 *
 * const result = await sharedMetadata({
 *  contract,
 * });
 *
 * ```
 */
async function sharedMetadata(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.sharedMetadata = sharedMetadata;
