// thirdweb/dist/cjs/wallets/in-app/implementations/lib/auth/abstract-login.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractLogin = void 0;
/**
 * @internal
 */
class AbstractLogin {
    /**
     * Used to manage the user's auth states. This should not be instantiated directly.
     * Call {@link InAppWalletSDK.auth} instead.
     * @internal
     */
    constructor({ baseUrl, querier, preLogin, postLogin, client, }) {
        Object.defineProperty(this, "LoginQuerier", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "preLogin", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "postLogin", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "client", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "baseUrl", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.baseUrl = baseUrl;
        this.LoginQuerier = querier;
        this.preLogin = preLogin;
        this.postLogin = postLogin;
        this.client = client;
    }
    /**
     * @internal
     */
    async sendEmailLoginOtp({ email, }) {
        await this.preLogin();
        const result = await this.LoginQuerier.call({
            procedureName: "sendThirdwebEmailLoginOtp",
            params: { email },
        });
        return result;
    }
    /**
     *
     * @internal
     */
    async sendSmsLoginOtp({ phoneNumber, }) {
        await this.preLogin();
        const result = await this.LoginQuerier.call({
            procedureName: "sendThirdwebSmsLoginOtp",
            params: { phoneNumber },
        });
        return result;
    }
}
exports.AbstractLogin = AbstractLogin;
