// thirdweb/dist/esm/react/web/ui/ConnectWallet/screens/StartScreen.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { keyframes } from "@emotion/react";
import { useConnectUI } from "../../../../core/hooks/others/useWalletConnectionCtx.js";
import { Img } from "../../components/Img.js";
import { Spacer } from "../../components/Spacer.js";
import { Container } from "../../components/basic.js";
import { Link } from "../../components/text.js";
import { Text } from "../../components/text.js";
import { useCustomTheme } from "../../design-system/CustomThemeProvider.js";
import { StyledDiv } from "../../design-system/elements.js";
import { spacing } from "../../design-system/index.js";
import { TOS } from "../Modal/TOS.js";
import { PoweredByThirdweb } from "../PoweredByTW.js";
import { GlobeIcon } from "../icons/GlobalIcon.js";
/**
 * @internal
 */
export function StartScreen() {
    const { connectLocale: locale, connectModal, client } = useConnectUI();
    const WelcomeScreen = connectModal.welcomeScreen;
    if (WelcomeScreen) {
        if (typeof WelcomeScreen === "function") {
            return _jsx(WelcomeScreen, {});
        }
    }
    const title = (typeof WelcomeScreen === "object" ? WelcomeScreen?.title : undefined) ||
        locale.welcomeScreen.defaultTitle;
    const subtitle = (typeof WelcomeScreen === "object" ? WelcomeScreen?.subtitle : undefined) ||
        locale.welcomeScreen.defaultSubtitle;
    const img = typeof WelcomeScreen === "object" ? WelcomeScreen?.img : undefined;
    const showTOS = connectModal.termsOfServiceUrl || connectModal.privacyPolicyUrl;
    return (_jsxs(Container, { fullHeight: true, animate: "fadein", flex: "column", children: [_jsxs(Container, { expand: true, flex: "column", center: "both", style: {
                    minHeight: "300px",
                }, p: "lg", children: [_jsx(Container, { flex: "row", center: "x", children: img ? (_jsx(Img, { src: img.src, width: img.width ? String(img.width) : undefined, height: img.height ? String(img.height) : undefined, client: client })) : (_jsx(GlobalContainer, { children: _jsx(GlobeIcon, { size: "150" }) })) }), _jsx(Spacer, { y: "xxl" }), _jsx(Text, { center: true, color: "primaryText", weight: 600, multiline: true, children: title }), _jsx(Spacer, { y: "md" }), _jsx(Text, { color: "secondaryText", multiline: true, style: {
                            textAlign: "center",
                        }, children: subtitle }), _jsx(Spacer, { y: "lg" }), _jsx(Link, { target: "_blank", center: true, href: "https://blog.thirdweb.com/web3-wallet/", children: locale.newToWallets })] }), _jsx(Container, { py: "lg", flex: "column", gap: "lg", children: _jsxs("div", { children: [showTOS && (_jsx(TOS, { termsOfServiceUrl: connectModal.termsOfServiceUrl, privacyPolicyUrl: connectModal.privacyPolicyUrl })), connectModal.showThirdwebBranding !== false && (_jsx(Container, { style: {
                                paddingTop: spacing.xl,
                            }, children: _jsx(PoweredByThirdweb, {}) }))] }) })] }));
}
const floatingAnimation = keyframes `
  from {
    transform: translateY(4px);
  }
  to {
    transform: translateY(-4px);
  }
`;
const GlobalContainer = /* @__PURE__ */ StyledDiv(() => {
    const theme = useCustomTheme();
    return {
        color: theme.colors.accentText,
        filter: `drop-shadow(0px 6px 10px ${theme.colors.accentText})`,
        animation: `${floatingAnimation} 2s ease infinite alternate`,
    };
});
