// thirdweb/dist/esm/react/web/ui/ConnectWallet/screens/ViewFunds.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useWalletBalance } from "../../../../core/hooks/others/useWalletBalance.js";
import { useActiveAccount, useActiveWalletChain, } from "../../../../core/hooks/wallets/wallet-hooks.js";
import { Skeleton } from "../../components/Skeleton.js";
import { Spacer } from "../../components/Spacer.js";
import { TokenIcon } from "../../components/TokenIcon.js";
import { Container, Line, ModalHeader } from "../../components/basic.js";
import { Text } from "../../components/text.js";
import { fontSize } from "../../design-system/index.js";
import { defaultTokens } from "../defaultTokens.js";
import { formatTokenBalance } from "./formatTokenBalance.js";
import { NATIVE_TOKEN, isNativeToken, } from "./nativeToken.js";
/**
 * @internal
 */
export function ViewFunds(props) {
    const activeChain = useActiveWalletChain();
    const supportedTokens = props.supportedTokens || defaultTokens;
    if (!activeChain) {
        return null;
    }
    const tokenList = (activeChain?.id ? supportedTokens[activeChain.id] : undefined) || [];
    return (_jsxs(Container, { style: {
            minHeight: "300px",
        }, children: [_jsx(Container, { p: "lg", children: _jsx(ModalHeader, { title: "View Funds", onBack: props.onBack }) }), _jsx(Line, {}), _jsxs(Container, { px: "sm", scrollY: true, style: {
                    maxHeight: "500px",
                }, children: [_jsx(Spacer, { y: "md" }), _jsx(TokenInfo, { token: NATIVE_TOKEN, chain: activeChain, client: props.client }), tokenList.map((token) => {
                        return (_jsx(TokenInfo, { token: token, chain: activeChain, client: props.client }, token.address));
                    }), _jsx(Spacer, { y: "lg" })] })] }));
}
function TokenInfo(props) {
    const account = useActiveAccount();
    const tokenBalanceQuery = useWalletBalance({
        address: account?.address,
        chain: props.chain,
        tokenAddress: isNativeToken(props.token) ? undefined : props.token.address,
        client: props.client,
    });
    const tokenName = isNativeToken(props.token)
        ? tokenBalanceQuery.data?.name
        : props.token.name;
    return (_jsxs(Container, { flex: "row", gap: "sm", p: "sm", children: [_jsx(TokenIcon, { token: props.token, chain: props.chain, size: "lg", client: props.client }), _jsxs(Container, { flex: "column", gap: "xxs", children: [tokenName ? (_jsx(Text, { size: "sm", color: "primaryText", children: tokenName })) : (_jsx(Skeleton, { height: fontSize.md, width: "150px" })), tokenBalanceQuery.data ? (_jsxs(Text, { size: "xs", children: [" ", formatTokenBalance(tokenBalanceQuery.data)] })) : (_jsx(Skeleton, { height: fontSize.xs, width: "100px" }))] })] }));
}
