// thirdweb/dist/cjs/extensions/erc4337/__generated__/IAccountPermissions/read/getAllActiveSigners.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllActiveSigners = exports.decodeGetAllActiveSignersResult = exports.isGetAllActiveSignersSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x8b52d723";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "tuple[]",
        name: "signers",
        components: [
            {
                type: "address",
                name: "signer",
            },
            {
                type: "address[]",
                name: "approvedTargets",
            },
            {
                type: "uint256",
                name: "nativeTokenLimitPerTransaction",
            },
            {
                type: "uint128",
                name: "startTimestamp",
            },
            {
                type: "uint128",
                name: "endTimestamp",
            },
        ],
    },
];
/**
 * Checks if the `getAllActiveSigners` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getAllActiveSigners` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isGetAllActiveSignersSupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = await isGetAllActiveSignersSupported(contract);
 * ```
 */
async function isGetAllActiveSignersSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetAllActiveSignersSupported = isGetAllActiveSignersSupported;
/**
 * Decodes the result of the getAllActiveSigners function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC4337
 * @example
 * ```ts
 * import { decodeGetAllActiveSignersResult } from "thirdweb/extensions/erc4337";
 * const result = decodeGetAllActiveSignersResult("...");
 * ```
 */
function decodeGetAllActiveSignersResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetAllActiveSignersResult = decodeGetAllActiveSignersResult;
/**
 * Calls the "getAllActiveSigners" function on the contract.
 * @param options - The options for the getAllActiveSigners function.
 * @returns The parsed result of the function call.
 * @extension ERC4337
 * @example
 * ```ts
 * import { getAllActiveSigners } from "thirdweb/extensions/erc4337";
 *
 * const result = await getAllActiveSigners({
 *  contract,
 * });
 *
 * ```
 */
async function getAllActiveSigners(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.getAllActiveSigners = getAllActiveSigners;
