// thirdweb/dist/cjs/extensions/erc4337/__generated__/IAccountPermissions/read/verifySignerPermissionRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifySignerPermissionRequest = exports.decodeVerifySignerPermissionRequestResult = exports.encodeVerifySignerPermissionRequest = exports.encodeVerifySignerPermissionRequestParams = exports.isVerifySignerPermissionRequestSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xa9082d84";
const FN_INPUTS = [
    {
        type: "tuple",
        name: "req",
        components: [
            {
                type: "address",
                name: "signer",
            },
            {
                type: "uint8",
                name: "isAdmin",
            },
            {
                type: "address[]",
                name: "approvedTargets",
            },
            {
                type: "uint256",
                name: "nativeTokenLimitPerTransaction",
            },
            {
                type: "uint128",
                name: "permissionStartTimestamp",
            },
            {
                type: "uint128",
                name: "permissionEndTimestamp",
            },
            {
                type: "uint128",
                name: "reqValidityStartTimestamp",
            },
            {
                type: "uint128",
                name: "reqValidityEndTimestamp",
            },
            {
                type: "bytes32",
                name: "uid",
            },
        ],
    },
    {
        type: "bytes",
        name: "signature",
    },
];
const FN_OUTPUTS = [
    {
        type: "bool",
        name: "success",
    },
    {
        type: "address",
        name: "signer",
    },
];
/**
 * Checks if the `verifySignerPermissionRequest` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `verifySignerPermissionRequest` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isVerifySignerPermissionRequestSupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = await isVerifySignerPermissionRequestSupported(contract);
 * ```
 */
async function isVerifySignerPermissionRequestSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isVerifySignerPermissionRequestSupported = isVerifySignerPermissionRequestSupported;
/**
 * Encodes the parameters for the "verifySignerPermissionRequest" function.
 * @param options - The options for the verifySignerPermissionRequest function.
 * @returns The encoded ABI parameters.
 * @extension ERC4337
 * @example
 * ```ts
 * import { encodeVerifySignerPermissionRequestParams } "thirdweb/extensions/erc4337";
 * const result = encodeVerifySignerPermissionRequestParams({
 *  req: ...,
 *  signature: ...,
 * });
 * ```
 */
function encodeVerifySignerPermissionRequestParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.req, options.signature]);
}
exports.encodeVerifySignerPermissionRequestParams = encodeVerifySignerPermissionRequestParams;
/**
 * Encodes the "verifySignerPermissionRequest" function into a Hex string with its parameters.
 * @param options - The options for the verifySignerPermissionRequest function.
 * @returns The encoded hexadecimal string.
 * @extension ERC4337
 * @example
 * ```ts
 * import { encodeVerifySignerPermissionRequest } "thirdweb/extensions/erc4337";
 * const result = encodeVerifySignerPermissionRequest({
 *  req: ...,
 *  signature: ...,
 * });
 * ```
 */
function encodeVerifySignerPermissionRequest(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeVerifySignerPermissionRequestParams(options).slice(2));
}
exports.encodeVerifySignerPermissionRequest = encodeVerifySignerPermissionRequest;
/**
 * Decodes the result of the verifySignerPermissionRequest function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC4337
 * @example
 * ```ts
 * import { decodeVerifySignerPermissionRequestResult } from "thirdweb/extensions/erc4337";
 * const result = decodeVerifySignerPermissionRequestResult("...");
 * ```
 */
function decodeVerifySignerPermissionRequestResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result);
}
exports.decodeVerifySignerPermissionRequestResult = decodeVerifySignerPermissionRequestResult;
/**
 * Calls the "verifySignerPermissionRequest" function on the contract.
 * @param options - The options for the verifySignerPermissionRequest function.
 * @returns The parsed result of the function call.
 * @extension ERC4337
 * @example
 * ```ts
 * import { verifySignerPermissionRequest } from "thirdweb/extensions/erc4337";
 *
 * const result = await verifySignerPermissionRequest({
 *  contract,
 *  req: ...,
 *  signature: ...,
 * });
 *
 * ```
 */
async function verifySignerPermissionRequest(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.req, options.signature],
    });
}
exports.verifySignerPermissionRequest = verifySignerPermissionRequest;
