// thirdweb/dist/esm/extensions/common/__generated__/IRoyalty/read/getDefaultRoyaltyInfo.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xb24f2d39";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "address",
    },
    {
        type: "uint16",
    },
];
/**
 * Checks if the `getDefaultRoyaltyInfo` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getDefaultRoyaltyInfo` method is supported.
 * @extension COMMON
 * @example
 * ```ts
 * import { isGetDefaultRoyaltyInfoSupported } from "thirdweb/extensions/common";
 *
 * const supported = await isGetDefaultRoyaltyInfoSupported(contract);
 * ```
 */
export async function isGetDefaultRoyaltyInfoSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getDefaultRoyaltyInfo function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension COMMON
 * @example
 * ```ts
 * import { decodeGetDefaultRoyaltyInfoResult } from "thirdweb/extensions/common";
 * const result = decodeGetDefaultRoyaltyInfoResult("...");
 * ```
 */
export function decodeGetDefaultRoyaltyInfoResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result);
}
/**
 * Calls the "getDefaultRoyaltyInfo" function on the contract.
 * @param options - The options for the getDefaultRoyaltyInfo function.
 * @returns The parsed result of the function call.
 * @extension COMMON
 * @example
 * ```ts
 * import { getDefaultRoyaltyInfo } from "thirdweb/extensions/common";
 *
 * const result = await getDefaultRoyaltyInfo({
 *  contract,
 * });
 *
 * ```
 */
export async function getDefaultRoyaltyInfo(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
