// thirdweb/dist/cjs/extensions/erc1155/__generated__/IERC1155Enumerable/read/nextTokenIdToMint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.nextTokenIdToMint = exports.decodeNextTokenIdToMintResult = exports.isNextTokenIdToMintSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x3b1475a7";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `nextTokenIdToMint` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `nextTokenIdToMint` method is supported.
 * @extension ERC1155
 * @example
 * ```ts
 * import { isNextTokenIdToMintSupported } from "thirdweb/extensions/erc1155";
 *
 * const supported = await isNextTokenIdToMintSupported(contract);
 * ```
 */
async function isNextTokenIdToMintSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isNextTokenIdToMintSupported = isNextTokenIdToMintSupported;
/**
 * Decodes the result of the nextTokenIdToMint function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC1155
 * @example
 * ```ts
 * import { decodeNextTokenIdToMintResult } from "thirdweb/extensions/erc1155";
 * const result = decodeNextTokenIdToMintResult("...");
 * ```
 */
function decodeNextTokenIdToMintResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeNextTokenIdToMintResult = decodeNextTokenIdToMintResult;
/**
 * Calls the "nextTokenIdToMint" function on the contract.
 * @param options - The options for the nextTokenIdToMint function.
 * @returns The parsed result of the function call.
 * @extension ERC1155
 * @example
 * ```ts
 * import { nextTokenIdToMint } from "thirdweb/extensions/erc1155";
 *
 * const result = await nextTokenIdToMint({
 *  contract,
 * });
 *
 * ```
 */
async function nextTokenIdToMint(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.nextTokenIdToMint = nextTokenIdToMint;
