// thirdweb/dist/cjs/extensions/erc4337/__generated__/IAccountPermissions/write/setPermissionsForSigner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setPermissionsForSigner = exports.encodeSetPermissionsForSigner = exports.encodeSetPermissionsForSignerParams = exports.isSetPermissionsForSignerSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x5892e236";
const FN_INPUTS = [
    {
        type: "tuple",
        name: "req",
        components: [
            {
                type: "address",
                name: "signer",
            },
            {
                type: "uint8",
                name: "isAdmin",
            },
            {
                type: "address[]",
                name: "approvedTargets",
            },
            {
                type: "uint256",
                name: "nativeTokenLimitPerTransaction",
            },
            {
                type: "uint128",
                name: "permissionStartTimestamp",
            },
            {
                type: "uint128",
                name: "permissionEndTimestamp",
            },
            {
                type: "uint128",
                name: "reqValidityStartTimestamp",
            },
            {
                type: "uint128",
                name: "reqValidityEndTimestamp",
            },
            {
                type: "bytes32",
                name: "uid",
            },
        ],
    },
    {
        type: "bytes",
        name: "signature",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `setPermissionsForSigner` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `setPermissionsForSigner` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isSetPermissionsForSignerSupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = await isSetPermissionsForSignerSupported(contract);
 * ```
 */
async function isSetPermissionsForSignerSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isSetPermissionsForSignerSupported = isSetPermissionsForSignerSupported;
/**
 * Encodes the parameters for the "setPermissionsForSigner" function.
 * @param options - The options for the setPermissionsForSigner function.
 * @returns The encoded ABI parameters.
 * @extension ERC4337
 * @example
 * ```ts
 * import { encodeSetPermissionsForSignerParams } "thirdweb/extensions/erc4337";
 * const result = encodeSetPermissionsForSignerParams({
 *  req: ...,
 *  signature: ...,
 * });
 * ```
 */
function encodeSetPermissionsForSignerParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.req, options.signature]);
}
exports.encodeSetPermissionsForSignerParams = encodeSetPermissionsForSignerParams;
/**
 * Encodes the "setPermissionsForSigner" function into a Hex string with its parameters.
 * @param options - The options for the setPermissionsForSigner function.
 * @returns The encoded hexadecimal string.
 * @extension ERC4337
 * @example
 * ```ts
 * import { encodeSetPermissionsForSigner } "thirdweb/extensions/erc4337";
 * const result = encodeSetPermissionsForSigner({
 *  req: ...,
 *  signature: ...,
 * });
 * ```
 */
function encodeSetPermissionsForSigner(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeSetPermissionsForSignerParams(options).slice(2));
}
exports.encodeSetPermissionsForSigner = encodeSetPermissionsForSigner;
/**
 * Prepares a transaction to call the "setPermissionsForSigner" function on the contract.
 * @param options - The options for the "setPermissionsForSigner" function.
 * @returns A prepared transaction object.
 * @extension ERC4337
 * @example
 * ```ts
 * import { setPermissionsForSigner } from "thirdweb/extensions/erc4337";
 *
 * const transaction = setPermissionsForSigner({
 *  contract,
 *  req: ...,
 *  signature: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function setPermissionsForSigner(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.req, resolvedOptions.signature];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.setPermissionsForSigner = setPermissionsForSigner;
