// thirdweb/dist/cjs/extensions/ens/__generated__/AddressResolver/read/name.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.name = exports.decodeNameResult = exports.encodeName = exports.encodeNameParams = exports.isNameSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x691f3431";
const FN_INPUTS = [
    {
        type: "bytes32",
        name: "name",
    },
];
const FN_OUTPUTS = [
    {
        type: "string",
    },
];
/**
 * Checks if the `name` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `name` method is supported.
 * @extension ENS
 * @example
 * ```ts
 * import { isNameSupported } from "thirdweb/extensions/ens";
 *
 * const supported = await isNameSupported(contract);
 * ```
 */
async function isNameSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isNameSupported = isNameSupported;
/**
 * Encodes the parameters for the "name" function.
 * @param options - The options for the name function.
 * @returns The encoded ABI parameters.
 * @extension ENS
 * @example
 * ```ts
 * import { encodeNameParams } "thirdweb/extensions/ens";
 * const result = encodeNameParams({
 *  name: ...,
 * });
 * ```
 */
function encodeNameParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.name]);
}
exports.encodeNameParams = encodeNameParams;
/**
 * Encodes the "name" function into a Hex string with its parameters.
 * @param options - The options for the name function.
 * @returns The encoded hexadecimal string.
 * @extension ENS
 * @example
 * ```ts
 * import { encodeName } "thirdweb/extensions/ens";
 * const result = encodeName({
 *  name: ...,
 * });
 * ```
 */
function encodeName(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeNameParams(options).slice(2));
}
exports.encodeName = encodeName;
/**
 * Decodes the result of the name function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ENS
 * @example
 * ```ts
 * import { decodeNameResult } from "thirdweb/extensions/ens";
 * const result = decodeNameResult("...");
 * ```
 */
function decodeNameResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeNameResult = decodeNameResult;
/**
 * Calls the "name" function on the contract.
 * @param options - The options for the name function.
 * @returns The parsed result of the function call.
 * @extension ENS
 * @example
 * ```ts
 * import { name } from "thirdweb/extensions/ens";
 *
 * const result = await name({
 *  contract,
 *  name: ...,
 * });
 *
 * ```
 */
async function name(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.name],
    });
}
exports.name = name;
