// thirdweb/dist/cjs/extensions/multicall3/__generated__/IMulticall3/write/aggregate3.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.aggregate3 = exports.encodeAggregate3 = exports.encodeAggregate3Params = exports.isAggregate3Supported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x82ad56cb";
const FN_INPUTS = [
    {
        components: [
            {
                internalType: "address",
                name: "target",
                type: "address",
            },
            {
                internalType: "bool",
                name: "allowFailure",
                type: "bool",
            },
            {
                internalType: "bytes",
                name: "callData",
                type: "bytes",
            },
        ],
        internalType: "struct Multicall3.Call3[]",
        name: "calls",
        type: "tuple[]",
    },
];
const FN_OUTPUTS = [
    {
        components: [
            {
                internalType: "bool",
                name: "success",
                type: "bool",
            },
            {
                internalType: "bytes",
                name: "returnData",
                type: "bytes",
            },
        ],
        internalType: "struct Multicall3.Result[]",
        name: "returnData",
        type: "tuple[]",
    },
];
/**
 * Checks if the `aggregate3` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `aggregate3` method is supported.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { isAggregate3Supported } from "thirdweb/extensions/multicall3";
 *
 * const supported = await isAggregate3Supported(contract);
 * ```
 */
async function isAggregate3Supported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isAggregate3Supported = isAggregate3Supported;
/**
 * Encodes the parameters for the "aggregate3" function.
 * @param options - The options for the aggregate3 function.
 * @returns The encoded ABI parameters.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { encodeAggregate3Params } "thirdweb/extensions/multicall3";
 * const result = encodeAggregate3Params({
 *  calls: ...,
 * });
 * ```
 */
function encodeAggregate3Params(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.calls]);
}
exports.encodeAggregate3Params = encodeAggregate3Params;
/**
 * Encodes the "aggregate3" function into a Hex string with its parameters.
 * @param options - The options for the aggregate3 function.
 * @returns The encoded hexadecimal string.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { encodeAggregate3 } "thirdweb/extensions/multicall3";
 * const result = encodeAggregate3({
 *  calls: ...,
 * });
 * ```
 */
function encodeAggregate3(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeAggregate3Params(options).slice(2));
}
exports.encodeAggregate3 = encodeAggregate3;
/**
 * Prepares a transaction to call the "aggregate3" function on the contract.
 * @param options - The options for the "aggregate3" function.
 * @returns A prepared transaction object.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { aggregate3 } from "thirdweb/extensions/multicall3";
 *
 * const transaction = aggregate3({
 *  contract,
 *  calls: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function aggregate3(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.calls];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.aggregate3 = aggregate3;
