// thirdweb/dist/cjs/extensions/multicall3/__generated__/IMulticall3/write/tryBlockAndAggregate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tryBlockAndAggregate = exports.encodeTryBlockAndAggregate = exports.encodeTryBlockAndAggregateParams = exports.isTryBlockAndAggregateSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x399542e9";
const FN_INPUTS = [
    {
        internalType: "bool",
        name: "requireSuccess",
        type: "bool",
    },
    {
        components: [
            {
                internalType: "address",
                name: "target",
                type: "address",
            },
            {
                internalType: "bytes",
                name: "callData",
                type: "bytes",
            },
        ],
        internalType: "struct Multicall3.Call[]",
        name: "calls",
        type: "tuple[]",
    },
];
const FN_OUTPUTS = [
    {
        internalType: "uint256",
        name: "blockNumber",
        type: "uint256",
    },
    {
        internalType: "bytes32",
        name: "blockHash",
        type: "bytes32",
    },
    {
        components: [
            {
                internalType: "bool",
                name: "success",
                type: "bool",
            },
            {
                internalType: "bytes",
                name: "returnData",
                type: "bytes",
            },
        ],
        internalType: "struct Multicall3.Result[]",
        name: "returnData",
        type: "tuple[]",
    },
];
/**
 * Checks if the `tryBlockAndAggregate` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `tryBlockAndAggregate` method is supported.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { isTryBlockAndAggregateSupported } from "thirdweb/extensions/multicall3";
 *
 * const supported = await isTryBlockAndAggregateSupported(contract);
 * ```
 */
async function isTryBlockAndAggregateSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isTryBlockAndAggregateSupported = isTryBlockAndAggregateSupported;
/**
 * Encodes the parameters for the "tryBlockAndAggregate" function.
 * @param options - The options for the tryBlockAndAggregate function.
 * @returns The encoded ABI parameters.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { encodeTryBlockAndAggregateParams } "thirdweb/extensions/multicall3";
 * const result = encodeTryBlockAndAggregateParams({
 *  requireSuccess: ...,
 *  calls: ...,
 * });
 * ```
 */
function encodeTryBlockAndAggregateParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [
        options.requireSuccess,
        options.calls,
    ]);
}
exports.encodeTryBlockAndAggregateParams = encodeTryBlockAndAggregateParams;
/**
 * Encodes the "tryBlockAndAggregate" function into a Hex string with its parameters.
 * @param options - The options for the tryBlockAndAggregate function.
 * @returns The encoded hexadecimal string.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { encodeTryBlockAndAggregate } "thirdweb/extensions/multicall3";
 * const result = encodeTryBlockAndAggregate({
 *  requireSuccess: ...,
 *  calls: ...,
 * });
 * ```
 */
function encodeTryBlockAndAggregate(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeTryBlockAndAggregateParams(options).slice(2));
}
exports.encodeTryBlockAndAggregate = encodeTryBlockAndAggregate;
/**
 * Prepares a transaction to call the "tryBlockAndAggregate" function on the contract.
 * @param options - The options for the "tryBlockAndAggregate" function.
 * @returns A prepared transaction object.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { tryBlockAndAggregate } from "thirdweb/extensions/multicall3";
 *
 * const transaction = tryBlockAndAggregate({
 *  contract,
 *  requireSuccess: ...,
 *  calls: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function tryBlockAndAggregate(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.requireSuccess, resolvedOptions.calls];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.tryBlockAndAggregate = tryBlockAndAggregate;
