// thirdweb/dist/cjs/extensions/marketplace/__generated__/IMarketplace/read/contractURI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.contractURI = exports.decodeContractURIResult = exports.isContractURISupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xe8a3d485";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "string",
    },
];
/**
 * Checks if the `contractURI` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `contractURI` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isContractURISupported } from "thirdweb/extensions/marketplace";
 *
 * const supported = await isContractURISupported(contract);
 * ```
 */
async function isContractURISupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isContractURISupported = isContractURISupported;
/**
 * Decodes the result of the contractURI function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { decodeContractURIResult } from "thirdweb/extensions/marketplace";
 * const result = decodeContractURIResult("...");
 * ```
 */
function decodeContractURIResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeContractURIResult = decodeContractURIResult;
/**
 * Calls the "contractURI" function on the contract.
 * @param options - The options for the contractURI function.
 * @returns The parsed result of the function call.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { contractURI } from "thirdweb/extensions/marketplace";
 *
 * const result = await contractURI({
 *  contract,
 * });
 *
 * ```
 */
async function contractURI(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.contractURI = contractURI;
