// thirdweb/dist/cjs/extensions/farcaster/__generated__/IKeyRegistry/read/gatewayFrozen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.gatewayFrozen = exports.decodeGatewayFrozenResult = exports.isGatewayFrozenSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x95e7549f";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "bool",
    },
];
/**
 * Checks if the `gatewayFrozen` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `gatewayFrozen` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isGatewayFrozenSupported } from "thirdweb/extensions/farcaster";
 *
 * const supported = await isGatewayFrozenSupported(contract);
 * ```
 */
async function isGatewayFrozenSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGatewayFrozenSupported = isGatewayFrozenSupported;
/**
 * Decodes the result of the gatewayFrozen function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { decodeGatewayFrozenResult } from "thirdweb/extensions/farcaster";
 * const result = decodeGatewayFrozenResult("...");
 * ```
 */
function decodeGatewayFrozenResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGatewayFrozenResult = decodeGatewayFrozenResult;
/**
 * Calls the "gatewayFrozen" function on the contract.
 * @param options - The options for the gatewayFrozen function.
 * @returns The parsed result of the function call.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { gatewayFrozen } from "thirdweb/extensions/farcaster";
 *
 * const result = await gatewayFrozen({
 *  contract,
 * });
 *
 * ```
 */
async function gatewayFrozen(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.gatewayFrozen = gatewayFrozen;
