// thirdweb/dist/cjs/extensions/erc721/drops/write/claimTo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.claimTo = void 0;
const get_claim_params_js_1 = require("../../../../utils/extensions/drops/get-claim-params.js");
const claim_js_1 = require("../../__generated__/IDrop/write/claim.js");
/**
 * Claim ERC721 NFTs to a specified address
 * @param options - The options for the transaction
 * @extension ERC721
 * @example
 * ```ts
 * import { claimTo } from "thirdweb/extensions/erc721";
 * const tx = await claimTo({
 *   contract,
 *   to: "0x...",
 *   quantity: 1n,
 * });
 * ```
 * @throws If no claim condition is set
 * @returns A promise that resolves with the submitted transaction hash.
 */
function claimTo(options) {
    return (0, claim_js_1.claim)({
        contract: options.contract,
        asyncParams: () => (0, get_claim_params_js_1.getClaimParams)({
            type: "erc721",
            contract: options.contract,
            to: options.to,
            quantity: options.quantity,
            from: options.from,
        }),
    });
}
exports.claimTo = claimTo;
