// thirdweb/dist/cjs/extensions/farcaster/__generated__/IIdRegistry/read/TRANSFER_TYPEHASH.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRANSFER_TYPEHASH = exports.decodeTRANSFER_TYPEHASHResult = exports.isTRANSFER_TYPEHASHSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x00bf26f4";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "bytes32",
    },
];
/**
 * Checks if the `TRANSFER_TYPEHASH` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `TRANSFER_TYPEHASH` method is supported.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { isTRANSFER_TYPEHASHSupported } from "thirdweb/extensions/farcaster";
 *
 * const supported = await isTRANSFER_TYPEHASHSupported(contract);
 * ```
 */
async function isTRANSFER_TYPEHASHSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isTRANSFER_TYPEHASHSupported = isTRANSFER_TYPEHASHSupported;
/**
 * Decodes the result of the TRANSFER_TYPEHASH function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { decodeTRANSFER_TYPEHASHResult } from "thirdweb/extensions/farcaster";
 * const result = decodeTRANSFER_TYPEHASHResult("...");
 * ```
 */
function decodeTRANSFER_TYPEHASHResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeTRANSFER_TYPEHASHResult = decodeTRANSFER_TYPEHASHResult;
/**
 * Calls the "TRANSFER_TYPEHASH" function on the contract.
 * @param options - The options for the TRANSFER_TYPEHASH function.
 * @returns The parsed result of the function call.
 * @extension FARCASTER
 * @example
 * ```ts
 * import { TRANSFER_TYPEHASH } from "thirdweb/extensions/farcaster";
 *
 * const result = await TRANSFER_TYPEHASH({
 *  contract,
 * });
 *
 * ```
 */
async function TRANSFER_TYPEHASH(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.TRANSFER_TYPEHASH = TRANSFER_TYPEHASH;
