// thirdweb/dist/esm/exports/extensions/common.js
// read
export { contractURI } from "../../extensions/common/__generated__/IContractMetadata/read/contractURI.js";
export { getContractMetadata } from "../../extensions/common/read/getContractMetadata.js";
export { owner } from "../../extensions/common/__generated__/IOwnable/read/owner.js";
export { name } from "../../extensions/common/read/name.js";
export { symbol } from "../../extensions/common/read/symbol.js";
// write
export { setContractURI, } from "../../extensions/common/__generated__/IContractMetadata/write/setContractURI.js";
export { setOwner, } from "../../extensions/common/__generated__/IOwnable/write/setOwner.js";
export { setContractMetadata, } from "../../extensions/common/write/setContractMetadata.js";
// events
export { ownerUpdatedEvent, } from "../../extensions/common/__generated__/IOwnable/events/OwnerUpdated.js";
// --------------------------------------------------------
// Royalty
// --------------------------------------------------------
// read
export { getDefaultRoyaltyInfo } from "../../extensions/common/__generated__/IRoyalty/read/getDefaultRoyaltyInfo.js";
export { getRoyaltyInfoForToken, } from "../../extensions/common/__generated__/IRoyalty/read/getRoyaltyInfoForToken.js";
// write
export { setDefaultRoyaltyInfo, } from "../../extensions/common/__generated__/IRoyalty/write/setDefaultRoyaltyInfo.js";
export { setRoyaltyInfoForToken, } from "../../extensions/common/__generated__/IRoyalty/write/setRoyaltyInfoForToken.js";
// --------------------------------------------------------
// Platform Fees
// --------------------------------------------------------
export { getPlatformFeeInfo } from "../../extensions/common/__generated__/IPlatformFee/read/getPlatformFeeInfo.js";
export { setPlatformFeeInfo, } from "../../extensions/common/__generated__/IPlatformFee/write/setPlatformFeeInfo.js";
// --------------------------------------------------------
// Primary Sale
// --------------------------------------------------------
export { primarySaleRecipient } from "../../extensions/common/__generated__/IPrimarySale/read/primarySaleRecipient.js";
export { setPrimarySaleRecipient, } from "../../extensions/common/__generated__/IPrimarySale/write/setPrimarySaleRecipient.js";
