// thirdweb/dist/cjs/extensions/marketplace/__generated__/IDirectListings/write/approveCurrencyForListing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.approveCurrencyForListing = exports.encodeApproveCurrencyForListing = exports.encodeApproveCurrencyForListingParams = exports.isApproveCurrencyForListingSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xea8f9a3c";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_listingId",
    },
    {
        type: "address",
        name: "_currency",
    },
    {
        type: "uint256",
        name: "_pricePerTokenInCurrency",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `approveCurrencyForListing` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `approveCurrencyForListing` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isApproveCurrencyForListingSupported } from "thirdweb/extensions/marketplace";
 *
 * const supported = await isApproveCurrencyForListingSupported(contract);
 * ```
 */
async function isApproveCurrencyForListingSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isApproveCurrencyForListingSupported = isApproveCurrencyForListingSupported;
/**
 * Encodes the parameters for the "approveCurrencyForListing" function.
 * @param options - The options for the approveCurrencyForListing function.
 * @returns The encoded ABI parameters.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { encodeApproveCurrencyForListingParams } "thirdweb/extensions/marketplace";
 * const result = encodeApproveCurrencyForListingParams({
 *  listingId: ...,
 *  currency: ...,
 *  pricePerTokenInCurrency: ...,
 * });
 * ```
 */
function encodeApproveCurrencyForListingParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [
        options.listingId,
        options.currency,
        options.pricePerTokenInCurrency,
    ]);
}
exports.encodeApproveCurrencyForListingParams = encodeApproveCurrencyForListingParams;
/**
 * Encodes the "approveCurrencyForListing" function into a Hex string with its parameters.
 * @param options - The options for the approveCurrencyForListing function.
 * @returns The encoded hexadecimal string.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { encodeApproveCurrencyForListing } "thirdweb/extensions/marketplace";
 * const result = encodeApproveCurrencyForListing({
 *  listingId: ...,
 *  currency: ...,
 *  pricePerTokenInCurrency: ...,
 * });
 * ```
 */
function encodeApproveCurrencyForListing(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeApproveCurrencyForListingParams(options).slice(2));
}
exports.encodeApproveCurrencyForListing = encodeApproveCurrencyForListing;
/**
 * Prepares a transaction to call the "approveCurrencyForListing" function on the contract.
 * @param options - The options for the "approveCurrencyForListing" function.
 * @returns A prepared transaction object.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { approveCurrencyForListing } from "thirdweb/extensions/marketplace";
 *
 * const transaction = approveCurrencyForListing({
 *  contract,
 *  listingId: ...,
 *  currency: ...,
 *  pricePerTokenInCurrency: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function approveCurrencyForListing(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [
                resolvedOptions.listingId,
                resolvedOptions.currency,
                resolvedOptions.pricePerTokenInCurrency,
            ];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.approveCurrencyForListing = approveCurrencyForListing;
