// thirdweb/dist/cjs/extensions/marketplace/__generated__/IDirectListings/write/createListing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createListing = exports.encodeCreateListing = exports.encodeCreateListingParams = exports.isCreateListingSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x746415b5";
const FN_INPUTS = [
    {
        type: "tuple",
        name: "_params",
        components: [
            {
                type: "address",
                name: "assetContract",
            },
            {
                type: "uint256",
                name: "tokenId",
            },
            {
                type: "uint256",
                name: "quantity",
            },
            {
                type: "address",
                name: "currency",
            },
            {
                type: "uint256",
                name: "pricePerToken",
            },
            {
                type: "uint128",
                name: "startTimestamp",
            },
            {
                type: "uint128",
                name: "endTimestamp",
            },
            {
                type: "bool",
                name: "reserved",
            },
        ],
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
        name: "listingId",
    },
];
/**
 * Checks if the `createListing` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `createListing` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isCreateListingSupported } from "thirdweb/extensions/marketplace";
 *
 * const supported = await isCreateListingSupported(contract);
 * ```
 */
async function isCreateListingSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isCreateListingSupported = isCreateListingSupported;
/**
 * Encodes the parameters for the "createListing" function.
 * @param options - The options for the createListing function.
 * @returns The encoded ABI parameters.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { encodeCreateListingParams } "thirdweb/extensions/marketplace";
 * const result = encodeCreateListingParams({
 *  params: ...,
 * });
 * ```
 */
function encodeCreateListingParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.params]);
}
exports.encodeCreateListingParams = encodeCreateListingParams;
/**
 * Encodes the "createListing" function into a Hex string with its parameters.
 * @param options - The options for the createListing function.
 * @returns The encoded hexadecimal string.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { encodeCreateListing } "thirdweb/extensions/marketplace";
 * const result = encodeCreateListing({
 *  params: ...,
 * });
 * ```
 */
function encodeCreateListing(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeCreateListingParams(options).slice(2));
}
exports.encodeCreateListing = encodeCreateListing;
/**
 * Prepares a transaction to call the "createListing" function on the contract.
 * @param options - The options for the "createListing" function.
 * @returns A prepared transaction object.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { createListing } from "thirdweb/extensions/marketplace";
 *
 * const transaction = createListing({
 *  contract,
 *  params: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function createListing(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.params];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.createListing = createListing;
