// thirdweb/dist/cjs/extensions/multicall3/__generated__/IMulticall3/read/getBlockNumber.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBlockNumber = exports.decodeGetBlockNumberResult = exports.isGetBlockNumberSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x42cbb15c";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        internalType: "uint256",
        name: "blockNumber",
        type: "uint256",
    },
];
/**
 * Checks if the `getBlockNumber` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getBlockNumber` method is supported.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { isGetBlockNumberSupported } from "thirdweb/extensions/multicall3";
 *
 * const supported = await isGetBlockNumberSupported(contract);
 * ```
 */
async function isGetBlockNumberSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetBlockNumberSupported = isGetBlockNumberSupported;
/**
 * Decodes the result of the getBlockNumber function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { decodeGetBlockNumberResult } from "thirdweb/extensions/multicall3";
 * const result = decodeGetBlockNumberResult("...");
 * ```
 */
function decodeGetBlockNumberResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetBlockNumberResult = decodeGetBlockNumberResult;
/**
 * Calls the "getBlockNumber" function on the contract.
 * @param options - The options for the getBlockNumber function.
 * @returns The parsed result of the function call.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { getBlockNumber } from "thirdweb/extensions/multicall3";
 *
 * const result = await getBlockNumber({
 *  contract,
 * });
 *
 * ```
 */
async function getBlockNumber(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.getBlockNumber = getBlockNumber;
