// thirdweb/dist/cjs/extensions/multicall3/__generated__/IMulticall3/read/getBasefee.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBasefee = exports.decodeGetBasefeeResult = exports.isGetBasefeeSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x3e64a696";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        internalType: "uint256",
        name: "basefee",
        type: "uint256",
    },
];
/**
 * Checks if the `getBasefee` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getBasefee` method is supported.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { isGetBasefeeSupported } from "thirdweb/extensions/multicall3";
 *
 * const supported = await isGetBasefeeSupported(contract);
 * ```
 */
async function isGetBasefeeSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetBasefeeSupported = isGetBasefeeSupported;
/**
 * Decodes the result of the getBasefee function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { decodeGetBasefeeResult } from "thirdweb/extensions/multicall3";
 * const result = decodeGetBasefeeResult("...");
 * ```
 */
function decodeGetBasefeeResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetBasefeeResult = decodeGetBasefeeResult;
/**
 * Calls the "getBasefee" function on the contract.
 * @param options - The options for the getBasefee function.
 * @returns The parsed result of the function call.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { getBasefee } from "thirdweb/extensions/multicall3";
 *
 * const result = await getBasefee({
 *  contract,
 * });
 *
 * ```
 */
async function getBasefee(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.getBasefee = getBasefee;
