// thirdweb/dist/cjs/extensions/multicall3/__generated__/IMulticall3/read/getCurrentBlockTimestamp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCurrentBlockTimestamp = exports.decodeGetCurrentBlockTimestampResult = exports.isGetCurrentBlockTimestampSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x0f28c97d";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        internalType: "uint256",
        name: "timestamp",
        type: "uint256",
    },
];
/**
 * Checks if the `getCurrentBlockTimestamp` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getCurrentBlockTimestamp` method is supported.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { isGetCurrentBlockTimestampSupported } from "thirdweb/extensions/multicall3";
 *
 * const supported = await isGetCurrentBlockTimestampSupported(contract);
 * ```
 */
async function isGetCurrentBlockTimestampSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetCurrentBlockTimestampSupported = isGetCurrentBlockTimestampSupported;
/**
 * Decodes the result of the getCurrentBlockTimestamp function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { decodeGetCurrentBlockTimestampResult } from "thirdweb/extensions/multicall3";
 * const result = decodeGetCurrentBlockTimestampResult("...");
 * ```
 */
function decodeGetCurrentBlockTimestampResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetCurrentBlockTimestampResult = decodeGetCurrentBlockTimestampResult;
/**
 * Calls the "getCurrentBlockTimestamp" function on the contract.
 * @param options - The options for the getCurrentBlockTimestamp function.
 * @returns The parsed result of the function call.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { getCurrentBlockTimestamp } from "thirdweb/extensions/multicall3";
 *
 * const result = await getCurrentBlockTimestamp({
 *  contract,
 * });
 *
 * ```
 */
async function getCurrentBlockTimestamp(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.getCurrentBlockTimestamp = getCurrentBlockTimestamp;
