// thirdweb/dist/cjs/extensions/multicall3/__generated__/IMulticall3/read/getCurrentBlockGasLimit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCurrentBlockGasLimit = exports.decodeGetCurrentBlockGasLimitResult = exports.isGetCurrentBlockGasLimitSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x86d516e8";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        internalType: "uint256",
        name: "gaslimit",
        type: "uint256",
    },
];
/**
 * Checks if the `getCurrentBlockGasLimit` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getCurrentBlockGasLimit` method is supported.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { isGetCurrentBlockGasLimitSupported } from "thirdweb/extensions/multicall3";
 *
 * const supported = await isGetCurrentBlockGasLimitSupported(contract);
 * ```
 */
async function isGetCurrentBlockGasLimitSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetCurrentBlockGasLimitSupported = isGetCurrentBlockGasLimitSupported;
/**
 * Decodes the result of the getCurrentBlockGasLimit function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { decodeGetCurrentBlockGasLimitResult } from "thirdweb/extensions/multicall3";
 * const result = decodeGetCurrentBlockGasLimitResult("...");
 * ```
 */
function decodeGetCurrentBlockGasLimitResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetCurrentBlockGasLimitResult = decodeGetCurrentBlockGasLimitResult;
/**
 * Calls the "getCurrentBlockGasLimit" function on the contract.
 * @param options - The options for the getCurrentBlockGasLimit function.
 * @returns The parsed result of the function call.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { getCurrentBlockGasLimit } from "thirdweb/extensions/multicall3";
 *
 * const result = await getCurrentBlockGasLimit({
 *  contract,
 * });
 *
 * ```
 */
async function getCurrentBlockGasLimit(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.getCurrentBlockGasLimit = getCurrentBlockGasLimit;
