// thirdweb/dist/cjs/extensions/multicall3/__generated__/IMulticall3/read/getEthBalance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEthBalance = exports.decodeGetEthBalanceResult = exports.encodeGetEthBalance = exports.encodeGetEthBalanceParams = exports.isGetEthBalanceSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x4d2301cc";
const FN_INPUTS = [
    {
        internalType: "address",
        name: "addr",
        type: "address",
    },
];
const FN_OUTPUTS = [
    {
        internalType: "uint256",
        name: "balance",
        type: "uint256",
    },
];
/**
 * Checks if the `getEthBalance` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getEthBalance` method is supported.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { isGetEthBalanceSupported } from "thirdweb/extensions/multicall3";
 *
 * const supported = await isGetEthBalanceSupported(contract);
 * ```
 */
async function isGetEthBalanceSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetEthBalanceSupported = isGetEthBalanceSupported;
/**
 * Encodes the parameters for the "getEthBalance" function.
 * @param options - The options for the getEthBalance function.
 * @returns The encoded ABI parameters.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { encodeGetEthBalanceParams } "thirdweb/extensions/multicall3";
 * const result = encodeGetEthBalanceParams({
 *  addr: ...,
 * });
 * ```
 */
function encodeGetEthBalanceParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.addr]);
}
exports.encodeGetEthBalanceParams = encodeGetEthBalanceParams;
/**
 * Encodes the "getEthBalance" function into a Hex string with its parameters.
 * @param options - The options for the getEthBalance function.
 * @returns The encoded hexadecimal string.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { encodeGetEthBalance } "thirdweb/extensions/multicall3";
 * const result = encodeGetEthBalance({
 *  addr: ...,
 * });
 * ```
 */
function encodeGetEthBalance(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetEthBalanceParams(options).slice(2));
}
exports.encodeGetEthBalance = encodeGetEthBalance;
/**
 * Decodes the result of the getEthBalance function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { decodeGetEthBalanceResult } from "thirdweb/extensions/multicall3";
 * const result = decodeGetEthBalanceResult("...");
 * ```
 */
function decodeGetEthBalanceResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetEthBalanceResult = decodeGetEthBalanceResult;
/**
 * Calls the "getEthBalance" function on the contract.
 * @param options - The options for the getEthBalance function.
 * @returns The parsed result of the function call.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { getEthBalance } from "thirdweb/extensions/multicall3";
 *
 * const result = await getEthBalance({
 *  contract,
 *  addr: ...,
 * });
 *
 * ```
 */
async function getEthBalance(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.addr],
    });
}
exports.getEthBalance = getEthBalance;
