// thirdweb/dist/cjs/extensions/multicall3/__generated__/IMulticall3/read/getLastBlockHash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLastBlockHash = exports.decodeGetLastBlockHashResult = exports.isGetLastBlockHashSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x27e86d6e";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        internalType: "bytes32",
        name: "blockHash",
        type: "bytes32",
    },
];
/**
 * Checks if the `getLastBlockHash` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getLastBlockHash` method is supported.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { isGetLastBlockHashSupported } from "thirdweb/extensions/multicall3";
 *
 * const supported = await isGetLastBlockHashSupported(contract);
 * ```
 */
async function isGetLastBlockHashSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetLastBlockHashSupported = isGetLastBlockHashSupported;
/**
 * Decodes the result of the getLastBlockHash function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { decodeGetLastBlockHashResult } from "thirdweb/extensions/multicall3";
 * const result = decodeGetLastBlockHashResult("...");
 * ```
 */
function decodeGetLastBlockHashResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetLastBlockHashResult = decodeGetLastBlockHashResult;
/**
 * Calls the "getLastBlockHash" function on the contract.
 * @param options - The options for the getLastBlockHash function.
 * @returns The parsed result of the function call.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { getLastBlockHash } from "thirdweb/extensions/multicall3";
 *
 * const result = await getLastBlockHash({
 *  contract,
 * });
 *
 * ```
 */
async function getLastBlockHash(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.getLastBlockHash = getLastBlockHash;
