// thirdweb/dist/cjs/extensions/erc4337/account/removeSessionKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeSessionKey = void 0;
const setPermissionsForSigner_js_1 = require("../__generated__/IAccountPermissions/write/setPermissionsForSigner.js");
const common_js_1 = require("./common.js");
/**
 * Removes session key permissions for a specified address.
 * @param options - The options for the removeSessionKey function.
 * @returns The transaction object to be sent.
 * @example
 * ```ts
 * import { removeSessionKey } from 'thirdweb/extensions/erc4337';
 *
 * const transaction = removeSessionKey({
 * contract,
 * account,
 * sessionKeyAddress
 * });
 * await sendTransaction({ transaction, account });
 * ```
 * @extension ERC4337
 */
function removeSessionKey(options) {
    const { contract, account, sessionKeyAddress } = options;
    return (0, setPermissionsForSigner_js_1.setPermissionsForSigner)({
        contract,
        async asyncParams() {
            const { req, signature } = await (0, common_js_1.signPermissionRequest)({
                account,
                contract,
                req: await (0, common_js_1.toContractPermissions)({
                    target: sessionKeyAddress,
                    permissions: {
                        approvedTargets: [],
                        nativeTokenLimitPerTransaction: 0,
                        permissionStartTimestamp: new Date(0),
                        permissionEndTimestamp: new Date(0),
                    },
                }),
            });
            return { signature, req };
        },
    });
}
exports.removeSessionKey = removeSessionKey;
