// thirdweb/dist/cjs/extensions/erc4626/__generated__/IERC4626/read/convertToAssets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertToAssets = exports.decodeConvertToAssetsResult = exports.encodeConvertToAssets = exports.encodeConvertToAssetsParams = exports.isConvertToAssetsSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x07a2d13a";
const FN_INPUTS = [
    {
        name: "shares",
        type: "uint256",
        internalType: "uint256",
    },
];
const FN_OUTPUTS = [
    {
        name: "assets",
        type: "uint256",
        internalType: "uint256",
    },
];
/**
 * Checks if the `convertToAssets` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `convertToAssets` method is supported.
 * @extension ERC4626
 * @example
 * ```ts
 * import { isConvertToAssetsSupported } from "thirdweb/extensions/erc4626";
 *
 * const supported = await isConvertToAssetsSupported(contract);
 * ```
 */
async function isConvertToAssetsSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isConvertToAssetsSupported = isConvertToAssetsSupported;
/**
 * Encodes the parameters for the "convertToAssets" function.
 * @param options - The options for the convertToAssets function.
 * @returns The encoded ABI parameters.
 * @extension ERC4626
 * @example
 * ```ts
 * import { encodeConvertToAssetsParams } "thirdweb/extensions/erc4626";
 * const result = encodeConvertToAssetsParams({
 *  shares: ...,
 * });
 * ```
 */
function encodeConvertToAssetsParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.shares]);
}
exports.encodeConvertToAssetsParams = encodeConvertToAssetsParams;
/**
 * Encodes the "convertToAssets" function into a Hex string with its parameters.
 * @param options - The options for the convertToAssets function.
 * @returns The encoded hexadecimal string.
 * @extension ERC4626
 * @example
 * ```ts
 * import { encodeConvertToAssets } "thirdweb/extensions/erc4626";
 * const result = encodeConvertToAssets({
 *  shares: ...,
 * });
 * ```
 */
function encodeConvertToAssets(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeConvertToAssetsParams(options).slice(2));
}
exports.encodeConvertToAssets = encodeConvertToAssets;
/**
 * Decodes the result of the convertToAssets function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC4626
 * @example
 * ```ts
 * import { decodeConvertToAssetsResult } from "thirdweb/extensions/erc4626";
 * const result = decodeConvertToAssetsResult("...");
 * ```
 */
function decodeConvertToAssetsResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeConvertToAssetsResult = decodeConvertToAssetsResult;
/**
 * Calls the "convertToAssets" function on the contract.
 * @param options - The options for the convertToAssets function.
 * @returns The parsed result of the function call.
 * @extension ERC4626
 * @example
 * ```ts
 * import { convertToAssets } from "thirdweb/extensions/erc4626";
 *
 * const result = await convertToAssets({
 *  contract,
 *  shares: ...,
 * });
 *
 * ```
 */
async function convertToAssets(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.shares],
    });
}
exports.convertToAssets = convertToAssets;
