// thirdweb/dist/esm/react/web/ui/components/Spinner.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { keyframes } from "@emotion/react";
import { useCustomTheme } from "../design-system/CustomThemeProvider.js";
import { StyledCircle, StyledSvg } from "../design-system/elements.js";
import { iconSize } from "../design-system/index.js";
/**
 * @internal
 */
export const Spinner = (props) => {
    const theme = useCustomTheme();
    return (_jsx(Svg, { style: {
            width: `${iconSize[props.size]}px`,
            height: `${iconSize[props.size]}px`,
        }, viewBox: "0 0 50 50", children: _jsx(Circle, { cx: "25", cy: "25", r: "20", fill: "none", stroke: theme.colors[props.color], strokeWidth: "4" }) }));
};
// animations
const dashAnimation = keyframes `
  0% {
    stroke-dasharray: 1, 150;
    stroke-dashoffset: 0;
  }
  50% {
    stroke-dasharray: 90, 150;
    stroke-dashoffset: -35;
  }
  100% {
    stroke-dasharray: 90, 150;
    stroke-dashoffset: -124;
  }
`;
const rotateAnimation = keyframes `
  100% {
    transform: rotate(360deg);
  }
`;
const Svg = /* @__PURE__ */ StyledSvg({
    animation: `${rotateAnimation} 2s linear infinite`,
    width: "1em",
    height: "1em",
});
const Circle = /* @__PURE__ */ StyledCircle({
    strokeLinecap: "round",
    animation: `${dashAnimation} 1.5s ease-in-out infinite`,
});
