// thirdweb/dist/cjs/extensions/thirdweb/__generated__/ITWMultichainRegistry/read/getMetadataUri.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMetadataUri = exports.decodeGetMetadataUriResult = exports.encodeGetMetadataUri = exports.encodeGetMetadataUriParams = exports.isGetMetadataUriSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xf4c2012d";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_chainId",
    },
    {
        type: "address",
        name: "_deployment",
    },
];
const FN_OUTPUTS = [
    {
        type: "string",
        name: "metadataUri",
    },
];
/**
 * Checks if the `getMetadataUri` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getMetadataUri` method is supported.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { isGetMetadataUriSupported } from "thirdweb/extensions/thirdweb";
 *
 * const supported = await isGetMetadataUriSupported(contract);
 * ```
 */
async function isGetMetadataUriSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetMetadataUriSupported = isGetMetadataUriSupported;
/**
 * Encodes the parameters for the "getMetadataUri" function.
 * @param options - The options for the getMetadataUri function.
 * @returns The encoded ABI parameters.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { encodeGetMetadataUriParams } "thirdweb/extensions/thirdweb";
 * const result = encodeGetMetadataUriParams({
 *  chainId: ...,
 *  deployment: ...,
 * });
 * ```
 */
function encodeGetMetadataUriParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.chainId, options.deployment]);
}
exports.encodeGetMetadataUriParams = encodeGetMetadataUriParams;
/**
 * Encodes the "getMetadataUri" function into a Hex string with its parameters.
 * @param options - The options for the getMetadataUri function.
 * @returns The encoded hexadecimal string.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { encodeGetMetadataUri } "thirdweb/extensions/thirdweb";
 * const result = encodeGetMetadataUri({
 *  chainId: ...,
 *  deployment: ...,
 * });
 * ```
 */
function encodeGetMetadataUri(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetMetadataUriParams(options).slice(2));
}
exports.encodeGetMetadataUri = encodeGetMetadataUri;
/**
 * Decodes the result of the getMetadataUri function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { decodeGetMetadataUriResult } from "thirdweb/extensions/thirdweb";
 * const result = decodeGetMetadataUriResult("...");
 * ```
 */
function decodeGetMetadataUriResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetMetadataUriResult = decodeGetMetadataUriResult;
/**
 * Calls the "getMetadataUri" function on the contract.
 * @param options - The options for the getMetadataUri function.
 * @returns The parsed result of the function call.
 * @extension THIRDWEB
 * @example
 * ```ts
 * import { getMetadataUri } from "thirdweb/extensions/thirdweb";
 *
 * const result = await getMetadataUri({
 *  contract,
 *  chainId: ...,
 *  deployment: ...,
 * });
 *
 * ```
 */
async function getMetadataUri(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.chainId, options.deployment],
    });
}
exports.getMetadataUri = getMetadataUri;
