// thirdweb/dist/cjs/extensions/erc6551/__generated__/IERC6551Account/read/token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.token = exports.decodeTokenResult = exports.isTokenSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xfc0c546a";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
        name: "chainId",
    },
    {
        type: "address",
        name: "tokenContract",
    },
    {
        type: "uint256",
        name: "tokenId",
    },
];
/**
 * Checks if the `token` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `token` method is supported.
 * @extension ERC6551
 * @example
 * ```ts
 * import { isTokenSupported } from "thirdweb/extensions/erc6551";
 *
 * const supported = await isTokenSupported(contract);
 * ```
 */
async function isTokenSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isTokenSupported = isTokenSupported;
/**
 * Decodes the result of the token function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC6551
 * @example
 * ```ts
 * import { decodeTokenResult } from "thirdweb/extensions/erc6551";
 * const result = decodeTokenResult("...");
 * ```
 */
function decodeTokenResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result);
}
exports.decodeTokenResult = decodeTokenResult;
/**
 * Calls the "token" function on the contract.
 * @param options - The options for the token function.
 * @returns The parsed result of the function call.
 * @extension ERC6551
 * @example
 * ```ts
 * import { token } from "thirdweb/extensions/erc6551";
 *
 * const result = await token({
 *  contract,
 * });
 *
 * ```
 */
async function token(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.token = token;
