// thirdweb/dist/esm/extensions/erc20/__generated__/DropSinglePhase/read/claimCondition.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xd637ed59";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        name: "startTimestamp",
        type: "uint256",
        internalType: "uint256",
    },
    {
        name: "maxClaimableSupply",
        type: "uint256",
        internalType: "uint256",
    },
    {
        name: "supplyClaimed",
        type: "uint256",
        internalType: "uint256",
    },
    {
        name: "quantityLimitPerWallet",
        type: "uint256",
        internalType: "uint256",
    },
    {
        name: "merkleRoot",
        type: "bytes32",
        internalType: "bytes32",
    },
    {
        name: "pricePerToken",
        type: "uint256",
        internalType: "uint256",
    },
    {
        name: "currency",
        type: "address",
        internalType: "address",
    },
    {
        name: "metadata",
        type: "string",
        internalType: "string",
    },
];
/**
 * Checks if the `claimCondition` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `claimCondition` method is supported.
 * @extension ERC20
 * @example
 * ```ts
 * import { isClaimConditionSupported } from "thirdweb/extensions/erc20";
 *
 * const supported = await isClaimConditionSupported(contract);
 * ```
 */
export async function isClaimConditionSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the claimCondition function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension ERC20
 * @example
 * ```ts
 * import { decodeClaimConditionResult } from "thirdweb/extensions/erc20";
 * const result = decodeClaimConditionResult("...");
 * ```
 */
export function decodeClaimConditionResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result);
}
/**
 * Calls the "claimCondition" function on the contract.
 * @param options - The options for the claimCondition function.
 * @returns The parsed result of the function call.
 * @extension ERC20
 * @example
 * ```ts
 * import { claimCondition } from "thirdweb/extensions/erc20";
 *
 * const result = await claimCondition({
 *  contract,
 * });
 *
 * ```
 */
export async function claimCondition(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
