// thirdweb/dist/cjs/extensions/erc1155/write/mintAdditionalSupplyTo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mintAdditionalSupplyTo = void 0;
const uri_js_1 = require("../__generated__/IERC1155/read/uri.js");
const mintTo_js_1 = require("../__generated__/IMintableERC1155/write/mintTo.js");
/**
 * Mints a "supply" number of additional ERC1155 tokens to the specified "to" address.
 * @param options - The transaction options.
 * @returns A promise that resolves to the transaction result.
 * @extension ERC1155
 * @example
 * ```ts
 * import { mintAdditionalSupplyTo } from "thirdweb/extensions/erc1155";
 * const transaction = mintAdditionalSupplyTo({
 *  contract,
 *  to: "0x...",
 *  tokenId: 1n,
 *  supply: 10n,
 * });
 *
 * const { transactionHash } = await sendTransaction({ transaction, account });
 *
 * ```
 */
function mintAdditionalSupplyTo(options) {
    return (0, mintTo_js_1.mintTo)({
        contract: options.contract,
        asyncParams: async () => {
            // we'll be re-using the exising token URI
            const tokenUri = await (0, uri_js_1.uri)({
                contract: options.contract,
                tokenId: options.tokenId,
            });
            return {
                to: options.to,
                tokenId: options.tokenId,
                uri: tokenUri,
                amount: options.supply,
            };
        },
    });
}
exports.mintAdditionalSupplyTo = mintAdditionalSupplyTo;
