// thirdweb/dist/cjs/extensions/common/__generated__/IRoyalty/read/getDefaultRoyaltyInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultRoyaltyInfo = exports.decodeGetDefaultRoyaltyInfoResult = exports.isGetDefaultRoyaltyInfoSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xb24f2d39";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "address",
    },
    {
        type: "uint16",
    },
];
/**
 * Checks if the `getDefaultRoyaltyInfo` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getDefaultRoyaltyInfo` method is supported.
 * @extension COMMON
 * @example
 * ```ts
 * import { isGetDefaultRoyaltyInfoSupported } from "thirdweb/extensions/common";
 *
 * const supported = await isGetDefaultRoyaltyInfoSupported(contract);
 * ```
 */
async function isGetDefaultRoyaltyInfoSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetDefaultRoyaltyInfoSupported = isGetDefaultRoyaltyInfoSupported;
/**
 * Decodes the result of the getDefaultRoyaltyInfo function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension COMMON
 * @example
 * ```ts
 * import { decodeGetDefaultRoyaltyInfoResult } from "thirdweb/extensions/common";
 * const result = decodeGetDefaultRoyaltyInfoResult("...");
 * ```
 */
function decodeGetDefaultRoyaltyInfoResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result);
}
exports.decodeGetDefaultRoyaltyInfoResult = decodeGetDefaultRoyaltyInfoResult;
/**
 * Calls the "getDefaultRoyaltyInfo" function on the contract.
 * @param options - The options for the getDefaultRoyaltyInfo function.
 * @returns The parsed result of the function call.
 * @extension COMMON
 * @example
 * ```ts
 * import { getDefaultRoyaltyInfo } from "thirdweb/extensions/common";
 *
 * const result = await getDefaultRoyaltyInfo({
 *  contract,
 * });
 *
 * ```
 */
async function getDefaultRoyaltyInfo(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.getDefaultRoyaltyInfo = getDefaultRoyaltyInfo;
