// thirdweb/dist/esm/wallets/in-app/implementations/constants/settings.js
/**
 * @internal
 */
export const IN_APP_WALLET_PATH = "/sdk/2022-08-12/embedded-wallet";
/**
 * @internal
 */
export const HEADLESS_GOOGLE_OAUTH_ROUTE = "/auth/headless-google-login-managed";
/**
 * @internal
 */
export const BASE_IN_APP_WALLET_URL = "embedded-wallet.thirdweb.com";
// STORAGE
/**
 * @internal
 */
export const WALLET_USER_DETAILS_LOCAL_STORAGE_NAME = (clientId) => `thirdwebEwsWalletUserDetails-${clientId}`;
/**
 * @internal
 */
export const WALLET_USER_ID_LOCAL_STORAGE_NAME = (clientId) => `thirdwebEwsWalletUserId-${clientId}`;
/**
 * @internal
 */
const AUTH_TOKEN_LOCAL_STORAGE_PREFIX = "walletToken";
/**
 * @internal
 */
export const AUTH_TOKEN_LOCAL_STORAGE_NAME = (clientId) => {
    return `${AUTH_TOKEN_LOCAL_STORAGE_PREFIX}-${clientId}`;
};
/**
 * @internal
 */
export const PASSKEY_CREDENTIAL_ID_LOCAL_STORAGE_NAME = (clientId) => {
    return `passkey-credential-id-${clientId}`;
};
/**
 * @internal
 */
const DEVICE_SHARE_LOCAL_STORAGE_PREFIX = "a";
/**
 * @internal
 */
export const DEVICE_SHARE_LOCAL_STORAGE_NAME = (clientId, userId) => `${DEVICE_SHARE_LOCAL_STORAGE_PREFIX}-${clientId}-${userId}`;
/**
 * @internal
 */
export const DEVICE_SHARE_LOCAL_STORAGE_NAME_DEPRECATED = (clientId) => `${DEVICE_SHARE_LOCAL_STORAGE_PREFIX}-${clientId}`;
