// thirdweb/dist/cjs/wallets/smart/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultPaymasterUrl = exports.getDefaultBundlerUrl = exports.MANAGED_ACCOUNT_GAS_BUFFER = exports.ENTRYPOINT_ADDRESS_v0_6 = exports.DEFAULT_ACCOUNT_FACTORY = exports.DUMMY_SIGNATURE = exports.DEBUG = void 0;
const domains_js_1 = require("../../../utils/domains.js");
// dev only
exports.DEBUG = false;
exports.DUMMY_SIGNATURE = "0xfffffffffffffffffffffffffffffff0000000000000000000000000000000007aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa1c";
exports.DEFAULT_ACCOUNT_FACTORY = "0x85e23b94e7F5E9cC1fF78BCe78cfb15B81f0DF00";
exports.ENTRYPOINT_ADDRESS_v0_6 = "0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789"; // v0.6
exports.MANAGED_ACCOUNT_GAS_BUFFER = 50000n;
/**
 * @internal
 */
const getDefaultBundlerUrl = (chain) => {
    const domain = (0, domains_js_1.getThirdwebDomains)().bundler;
    return `https://${chain.id}.${domain}/`;
};
exports.getDefaultBundlerUrl = getDefaultBundlerUrl;
/**
 * @internal
 */
const getDefaultPaymasterUrl = (chain) => {
    const domain = (0, domains_js_1.getThirdwebDomains)().bundler;
    return `https://${chain.id}.${domain}/`;
};
exports.getDefaultPaymasterUrl = getDefaultPaymasterUrl;
