// thirdweb/dist/esm/extensions/multicall3/__generated__/IMulticall3/read/getChainId.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x3408e470";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        internalType: "uint256",
        name: "chainid",
        type: "uint256",
    },
];
/**
 * Checks if the `getChainId` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getChainId` method is supported.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { isGetChainIdSupported } from "thirdweb/extensions/multicall3";
 *
 * const supported = await isGetChainIdSupported(contract);
 * ```
 */
export async function isGetChainIdSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the getChainId function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { decodeGetChainIdResult } from "thirdweb/extensions/multicall3";
 * const result = decodeGetChainIdResult("...");
 * ```
 */
export function decodeGetChainIdResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getChainId" function on the contract.
 * @param options - The options for the getChainId function.
 * @returns The parsed result of the function call.
 * @extension MULTICALL3
 * @example
 * ```ts
 * import { getChainId } from "thirdweb/extensions/multicall3";
 *
 * const result = await getChainId({
 *  contract,
 * });
 *
 * ```
 */
export async function getChainId(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
