// thirdweb/dist/cjs/extensions/erc20/drops/write/claimTo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.claimTo = void 0;
const get_claim_params_js_1 = require("../../../../utils/extensions/drops/get-claim-params.js");
const claim_js_1 = require("../../__generated__/IDropERC20/write/claim.js");
const decimals_js_1 = require("../../read/decimals.js");
/**
 * Claim ERC20 NFTs to a specified address
 * @param options - The options for the transaction
 * @extension ERC20
 * @example
 * ```ts
 * import { claimTo } from "thirdweb/extensions/erc20";
 * const tx = await claimTo({
 *   contract,
 *   to: "0x...",
 *   quantity: 100n,
 * });
 * ```
 * @throws If no claim condition is set
 * @returns A promise that resolves with the submitted transaction hash.
 */
function claimTo(options) {
    return (0, claim_js_1.claim)({
        contract: options.contract,
        asyncParams: async () => {
            const quantity = await (async () => {
                if ("quantityInWei" in options) {
                    return options.quantityInWei;
                }
                const { toUnits } = await Promise.resolve().then(() => require("../../../../utils/units.js"));
                return toUnits(options.quantity, await (0, decimals_js_1.decimals)({ contract: options.contract }));
            })();
            return (0, get_claim_params_js_1.getClaimParams)({
                type: "erc20",
                contract: options.contract,
                to: options.to,
                quantity,
                from: options.from,
                tokenDecimals: await (0, decimals_js_1.decimals)({ contract: options.contract }),
            });
        },
    });
}
exports.claimTo = claimTo;
