// thirdweb/dist/esm/extensions/marketplace/__generated__/IMarketplace/read/contractType.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0xcb2ef6f7";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "bytes32",
    },
];
/**
 * Checks if the `contractType` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `contractType` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isContractTypeSupported } from "thirdweb/extensions/marketplace";
 *
 * const supported = await isContractTypeSupported(contract);
 * ```
 */
export async function isContractTypeSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Decodes the result of the contractType function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { decodeContractTypeResult } from "thirdweb/extensions/marketplace";
 * const result = decodeContractTypeResult("...");
 * ```
 */
export function decodeContractTypeResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "contractType" function on the contract.
 * @param options - The options for the contractType function.
 * @returns The parsed result of the function call.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { contractType } from "thirdweb/extensions/marketplace";
 *
 * const result = await contractType({
 *  contract,
 * });
 *
 * ```
 */
export async function contractType(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
