// thirdweb/dist/cjs/extensions/erc1155/__generated__/IERC1155Receiver/write/onERC1155BatchReceived.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.onERC1155BatchReceived = exports.encodeOnERC1155BatchReceived = exports.encodeOnERC1155BatchReceivedParams = exports.isOnERC1155BatchReceivedSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xbc197c81";
const FN_INPUTS = [
    {
        type: "address",
        name: "operator",
    },
    {
        type: "address",
        name: "from",
    },
    {
        type: "uint256[]",
        name: "ids",
    },
    {
        type: "uint256[]",
        name: "values",
    },
    {
        type: "bytes",
        name: "data",
    },
];
const FN_OUTPUTS = [
    {
        type: "bytes4",
    },
];
/**
 * Checks if the `onERC1155BatchReceived` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `onERC1155BatchReceived` method is supported.
 * @extension ERC1155
 * @example
 * ```ts
 * import { isOnERC1155BatchReceivedSupported } from "thirdweb/extensions/erc1155";
 *
 * const supported = await isOnERC1155BatchReceivedSupported(contract);
 * ```
 */
async function isOnERC1155BatchReceivedSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isOnERC1155BatchReceivedSupported = isOnERC1155BatchReceivedSupported;
/**
 * Encodes the parameters for the "onERC1155BatchReceived" function.
 * @param options - The options for the onERC1155BatchReceived function.
 * @returns The encoded ABI parameters.
 * @extension ERC1155
 * @example
 * ```ts
 * import { encodeOnERC1155BatchReceivedParams } "thirdweb/extensions/erc1155";
 * const result = encodeOnERC1155BatchReceivedParams({
 *  operator: ...,
 *  from: ...,
 *  ids: ...,
 *  values: ...,
 *  data: ...,
 * });
 * ```
 */
function encodeOnERC1155BatchReceivedParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [
        options.operator,
        options.from,
        options.ids,
        options.values,
        options.data,
    ]);
}
exports.encodeOnERC1155BatchReceivedParams = encodeOnERC1155BatchReceivedParams;
/**
 * Encodes the "onERC1155BatchReceived" function into a Hex string with its parameters.
 * @param options - The options for the onERC1155BatchReceived function.
 * @returns The encoded hexadecimal string.
 * @extension ERC1155
 * @example
 * ```ts
 * import { encodeOnERC1155BatchReceived } "thirdweb/extensions/erc1155";
 * const result = encodeOnERC1155BatchReceived({
 *  operator: ...,
 *  from: ...,
 *  ids: ...,
 *  values: ...,
 *  data: ...,
 * });
 * ```
 */
function encodeOnERC1155BatchReceived(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeOnERC1155BatchReceivedParams(options).slice(2));
}
exports.encodeOnERC1155BatchReceived = encodeOnERC1155BatchReceived;
/**
 * Prepares a transaction to call the "onERC1155BatchReceived" function on the contract.
 * @param options - The options for the "onERC1155BatchReceived" function.
 * @returns A prepared transaction object.
 * @extension ERC1155
 * @example
 * ```ts
 * import { onERC1155BatchReceived } from "thirdweb/extensions/erc1155";
 *
 * const transaction = onERC1155BatchReceived({
 *  contract,
 *  operator: ...,
 *  from: ...,
 *  ids: ...,
 *  values: ...,
 *  data: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function onERC1155BatchReceived(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [
                resolvedOptions.operator,
                resolvedOptions.from,
                resolvedOptions.ids,
                resolvedOptions.values,
                resolvedOptions.data,
            ];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.onERC1155BatchReceived = onERC1155BatchReceived;
