// thirdweb/dist/cjs/extensions/erc4337/__generated__/IEntryPoint/write/simulateValidation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.simulateValidation = exports.encodeSimulateValidation = exports.encodeSimulateValidationParams = exports.isSimulateValidationSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xee219423";
const FN_INPUTS = [
    {
        type: "tuple",
        name: "userOp",
        components: [
            {
                type: "address",
                name: "sender",
            },
            {
                type: "uint256",
                name: "nonce",
            },
            {
                type: "bytes",
                name: "initCode",
            },
            {
                type: "bytes",
                name: "callData",
            },
            {
                type: "uint256",
                name: "callGasLimit",
            },
            {
                type: "uint256",
                name: "verificationGasLimit",
            },
            {
                type: "uint256",
                name: "preVerificationGas",
            },
            {
                type: "uint256",
                name: "maxFeePerGas",
            },
            {
                type: "uint256",
                name: "maxPriorityFeePerGas",
            },
            {
                type: "bytes",
                name: "paymasterAndData",
            },
            {
                type: "bytes",
                name: "signature",
            },
        ],
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `simulateValidation` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `simulateValidation` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isSimulateValidationSupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = await isSimulateValidationSupported(contract);
 * ```
 */
async function isSimulateValidationSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isSimulateValidationSupported = isSimulateValidationSupported;
/**
 * Encodes the parameters for the "simulateValidation" function.
 * @param options - The options for the simulateValidation function.
 * @returns The encoded ABI parameters.
 * @extension ERC4337
 * @example
 * ```ts
 * import { encodeSimulateValidationParams } "thirdweb/extensions/erc4337";
 * const result = encodeSimulateValidationParams({
 *  userOp: ...,
 * });
 * ```
 */
function encodeSimulateValidationParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.userOp]);
}
exports.encodeSimulateValidationParams = encodeSimulateValidationParams;
/**
 * Encodes the "simulateValidation" function into a Hex string with its parameters.
 * @param options - The options for the simulateValidation function.
 * @returns The encoded hexadecimal string.
 * @extension ERC4337
 * @example
 * ```ts
 * import { encodeSimulateValidation } "thirdweb/extensions/erc4337";
 * const result = encodeSimulateValidation({
 *  userOp: ...,
 * });
 * ```
 */
function encodeSimulateValidation(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeSimulateValidationParams(options).slice(2));
}
exports.encodeSimulateValidation = encodeSimulateValidation;
/**
 * Prepares a transaction to call the "simulateValidation" function on the contract.
 * @param options - The options for the "simulateValidation" function.
 * @returns A prepared transaction object.
 * @extension ERC4337
 * @example
 * ```ts
 * import { simulateValidation } from "thirdweb/extensions/erc4337";
 *
 * const transaction = simulateValidation({
 *  contract,
 *  userOp: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function simulateValidation(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.userOp];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.simulateValidation = simulateValidation;
