// thirdweb/dist/cjs/extensions/erc4337/__generated__/IEntryPoint/write/unlockStake.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unlockStake = exports.isUnlockStakeSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xbb9fe6bf";
const FN_INPUTS = [];
const FN_OUTPUTS = [];
/**
 * Checks if the `unlockStake` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `unlockStake` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isUnlockStakeSupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = await isUnlockStakeSupported(contract);
 * ```
 */
async function isUnlockStakeSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isUnlockStakeSupported = isUnlockStakeSupported;
/**
 * Prepares a transaction to call the "unlockStake" function on the contract.
 * @param options - The options for the "unlockStake" function.
 * @returns A prepared transaction object.
 * @extension ERC4337
 * @example
 * ```ts
 * import { unlockStake } from "thirdweb/extensions/erc4337";
 *
 * const transaction = unlockStake();
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function unlockStake(options) {
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.unlockStake = unlockStake;
