// thirdweb/dist/cjs/extensions/erc4337/__generated__/IEntryPoint/write/withdrawStake.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withdrawStake = exports.encodeWithdrawStake = exports.encodeWithdrawStakeParams = exports.isWithdrawStakeSupported = exports.FN_SELECTOR = void 0;
const prepare_contract_call_js_1 = require("../../../../../transaction/prepare-contract-call.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const once_js_1 = require("../../../../../utils/promise/once.js");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xc23a5cea";
const FN_INPUTS = [
    {
        type: "address",
        name: "withdrawAddress",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `withdrawStake` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `withdrawStake` method is supported.
 * @extension ERC4337
 * @example
 * ```ts
 * import { isWithdrawStakeSupported } from "thirdweb/extensions/erc4337";
 *
 * const supported = await isWithdrawStakeSupported(contract);
 * ```
 */
async function isWithdrawStakeSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isWithdrawStakeSupported = isWithdrawStakeSupported;
/**
 * Encodes the parameters for the "withdrawStake" function.
 * @param options - The options for the withdrawStake function.
 * @returns The encoded ABI parameters.
 * @extension ERC4337
 * @example
 * ```ts
 * import { encodeWithdrawStakeParams } "thirdweb/extensions/erc4337";
 * const result = encodeWithdrawStakeParams({
 *  withdrawAddress: ...,
 * });
 * ```
 */
function encodeWithdrawStakeParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.withdrawAddress]);
}
exports.encodeWithdrawStakeParams = encodeWithdrawStakeParams;
/**
 * Encodes the "withdrawStake" function into a Hex string with its parameters.
 * @param options - The options for the withdrawStake function.
 * @returns The encoded hexadecimal string.
 * @extension ERC4337
 * @example
 * ```ts
 * import { encodeWithdrawStake } "thirdweb/extensions/erc4337";
 * const result = encodeWithdrawStake({
 *  withdrawAddress: ...,
 * });
 * ```
 */
function encodeWithdrawStake(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeWithdrawStakeParams(options).slice(2));
}
exports.encodeWithdrawStake = encodeWithdrawStake;
/**
 * Prepares a transaction to call the "withdrawStake" function on the contract.
 * @param options - The options for the "withdrawStake" function.
 * @returns A prepared transaction object.
 * @extension ERC4337
 * @example
 * ```ts
 * import { withdrawStake } from "thirdweb/extensions/erc4337";
 *
 * const transaction = withdrawStake({
 *  contract,
 *  withdrawAddress: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
function withdrawStake(options) {
    const asyncOptions = (0, once_js_1.once)(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return (0, prepare_contract_call_js_1.prepareContractCall)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.withdrawAddress];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
exports.withdrawStake = withdrawStake;
