// thirdweb/dist/cjs/extensions/marketplace/__generated__/IOffers/read/getAllOffers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllOffers = exports.decodeGetAllOffersResult = exports.encodeGetAllOffers = exports.encodeGetAllOffersParams = exports.isGetAllOffersSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xc1edcfbe";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_startId",
    },
    {
        type: "uint256",
        name: "_endId",
    },
];
const FN_OUTPUTS = [
    {
        type: "tuple[]",
        name: "offers",
        components: [
            {
                type: "uint256",
                name: "offerId",
            },
            {
                type: "uint256",
                name: "tokenId",
            },
            {
                type: "uint256",
                name: "quantity",
            },
            {
                type: "uint256",
                name: "totalPrice",
            },
            {
                type: "uint256",
                name: "expirationTimestamp",
            },
            {
                type: "address",
                name: "offeror",
            },
            {
                type: "address",
                name: "assetContract",
            },
            {
                type: "address",
                name: "currency",
            },
            {
                type: "uint8",
                name: "tokenType",
            },
            {
                type: "uint8",
                name: "status",
            },
        ],
    },
];
/**
 * Checks if the `getAllOffers` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `getAllOffers` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isGetAllOffersSupported } from "thirdweb/extensions/marketplace";
 *
 * const supported = await isGetAllOffersSupported(contract);
 * ```
 */
async function isGetAllOffersSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isGetAllOffersSupported = isGetAllOffersSupported;
/**
 * Encodes the parameters for the "getAllOffers" function.
 * @param options - The options for the getAllOffers function.
 * @returns The encoded ABI parameters.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { encodeGetAllOffersParams } "thirdweb/extensions/marketplace";
 * const result = encodeGetAllOffersParams({
 *  startId: ...,
 *  endId: ...,
 * });
 * ```
 */
function encodeGetAllOffersParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.startId, options.endId]);
}
exports.encodeGetAllOffersParams = encodeGetAllOffersParams;
/**
 * Encodes the "getAllOffers" function into a Hex string with its parameters.
 * @param options - The options for the getAllOffers function.
 * @returns The encoded hexadecimal string.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { encodeGetAllOffers } "thirdweb/extensions/marketplace";
 * const result = encodeGetAllOffers({
 *  startId: ...,
 *  endId: ...,
 * });
 * ```
 */
function encodeGetAllOffers(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetAllOffersParams(options).slice(2));
}
exports.encodeGetAllOffers = encodeGetAllOffers;
/**
 * Decodes the result of the getAllOffers function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { decodeGetAllOffersResult } from "thirdweb/extensions/marketplace";
 * const result = decodeGetAllOffersResult("...");
 * ```
 */
function decodeGetAllOffersResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeGetAllOffersResult = decodeGetAllOffersResult;
/**
 * Calls the "getAllOffers" function on the contract.
 * @param options - The options for the getAllOffers function.
 * @returns The parsed result of the function call.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { getAllOffers } from "thirdweb/extensions/marketplace";
 *
 * const result = await getAllOffers({
 *  contract,
 *  startId: ...,
 *  endId: ...,
 * });
 *
 * ```
 */
async function getAllOffers(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.startId, options.endId],
    });
}
exports.getAllOffers = getAllOffers;
