// thirdweb/dist/cjs/extensions/marketplace/__generated__/IOffers/read/totalOffers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.totalOffers = exports.decodeTotalOffersResult = exports.isTotalOffersSupported = exports.FN_SELECTOR = void 0;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xa9fd8ed1";
const FN_INPUTS = [];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `totalOffers` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `totalOffers` method is supported.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { isTotalOffersSupported } from "thirdweb/extensions/marketplace";
 *
 * const supported = await isTotalOffersSupported(contract);
 * ```
 */
async function isTotalOffersSupported(contract) {
    return (0, detectExtension_js_1.detectMethod)({
        contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
exports.isTotalOffersSupported = isTotalOffersSupported;
/**
 * Decodes the result of the totalOffers function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { decodeTotalOffersResult } from "thirdweb/extensions/marketplace";
 * const result = decodeTotalOffersResult("...");
 * ```
 */
function decodeTotalOffersResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
exports.decodeTotalOffersResult = decodeTotalOffersResult;
/**
 * Calls the "totalOffers" function on the contract.
 * @param options - The options for the totalOffers function.
 * @returns The parsed result of the function call.
 * @extension MARKETPLACE
 * @example
 * ```ts
 * import { totalOffers } from "thirdweb/extensions/marketplace";
 *
 * const result = await totalOffers({
 *  contract,
 * });
 *
 * ```
 */
async function totalOffers(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [],
    });
}
exports.totalOffers = totalOffers;
