// thirdweb/dist/cjs/exports/extensions/permissions.js
"use strict";
//--------------------------------------------------
// Permissions
// --------------------------------------------------------
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllRoleMembers = exports.getRoleMemberCount = exports.getRoleMember = exports.roleAdminChangedEvent = exports.roleRevokedEvent = exports.roleGrantedEvent = exports.renounceRole = exports.revokeRole = exports.grantRole = exports.getRoleAdmin = exports.hasRole = void 0;
// READ
var hasRole_js_1 = require("../../extensions/permissions/read/hasRole.js");
Object.defineProperty(exports, "hasRole", { enumerable: true, get: function () { return hasRole_js_1.hasRole; } });
var getRoleAdmin_js_1 = require("../../extensions/permissions/read/getRoleAdmin.js");
Object.defineProperty(exports, "getRoleAdmin", { enumerable: true, get: function () { return getRoleAdmin_js_1.getRoleAdmin; } });
// WRITE
var grant_js_1 = require("../../extensions/permissions/write/grant.js");
Object.defineProperty(exports, "grantRole", { enumerable: true, get: function () { return grant_js_1.grantRole; } });
var revokeRole_js_1 = require("../../extensions/permissions/write/revokeRole.js");
Object.defineProperty(exports, "revokeRole", { enumerable: true, get: function () { return revokeRole_js_1.revokeRole; } });
var renounceRole_js_1 = require("../../extensions/permissions/write/renounceRole.js");
Object.defineProperty(exports, "renounceRole", { enumerable: true, get: function () { return renounceRole_js_1.renounceRole; } });
// EVENTS
var RoleGranted_js_1 = require("../../extensions/permissions/__generated__/IPermissions/events/RoleGranted.js");
Object.defineProperty(exports, "roleGrantedEvent", { enumerable: true, get: function () { return RoleGranted_js_1.roleGrantedEvent; } });
var RoleRevoked_js_1 = require("../../extensions/permissions/__generated__/IPermissions/events/RoleRevoked.js");
Object.defineProperty(exports, "roleRevokedEvent", { enumerable: true, get: function () { return RoleRevoked_js_1.roleRevokedEvent; } });
var RoleAdminChanged_js_1 = require("../../extensions/permissions/__generated__/IPermissions/events/RoleAdminChanged.js");
Object.defineProperty(exports, "roleAdminChangedEvent", { enumerable: true, get: function () { return RoleAdminChanged_js_1.roleAdminChangedEvent; } });
// --------------------------------------------------------
// PermissionsEnumerable
// --------------------------------------------------------
// READ
var getRoleMember_js_1 = require("../../extensions/permissions/read/getRoleMember.js");
Object.defineProperty(exports, "getRoleMember", { enumerable: true, get: function () { return getRoleMember_js_1.getRoleMember; } });
var getRoleMemberCount_js_1 = require("../../extensions/permissions/read/getRoleMemberCount.js");
Object.defineProperty(exports, "getRoleMemberCount", { enumerable: true, get: function () { return getRoleMemberCount_js_1.getRoleMemberCount; } });
var getAllMembers_js_1 = require("../../extensions/permissions/read/getAllMembers.js");
Object.defineProperty(exports, "getAllRoleMembers", { enumerable: true, get: function () { return getAllMembers_js_1.getAllRoleMembers; } });
