// thirdweb/dist/esm/extensions/common/__generated__/IRoyaltyPayments/write/setRoyaltyEngine.js
import { prepareContractCall } from "../../../../../transaction/prepare-contract-call.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { once } from "../../../../../utils/promise/once.js";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x21ede032";
const FN_INPUTS = [
    {
        type: "address",
        name: "_royaltyEngineAddress",
    },
];
const FN_OUTPUTS = [];
/**
 * Checks if the `setRoyaltyEngine` method is supported by the given contract.
 * @param contract The ThirdwebContract.
 * @returns A promise that resolves to a boolean indicating if the `setRoyaltyEngine` method is supported.
 * @extension COMMON
 * @example
 * ```ts
 * import { isSetRoyaltyEngineSupported } from "thirdweb/extensions/common";
 *
 * const supported = await isSetRoyaltyEngineSupported(contract);
 * ```
 */
export async function isSetRoyaltyEngineSupported(contract) {
    return detectMethod({
        contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "setRoyaltyEngine" function.
 * @param options - The options for the setRoyaltyEngine function.
 * @returns The encoded ABI parameters.
 * @extension COMMON
 * @example
 * ```ts
 * import { encodeSetRoyaltyEngineParams } "thirdweb/extensions/common";
 * const result = encodeSetRoyaltyEngineParams({
 *  royaltyEngineAddress: ...,
 * });
 * ```
 */
export function encodeSetRoyaltyEngineParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.royaltyEngineAddress]);
}
/**
 * Encodes the "setRoyaltyEngine" function into a Hex string with its parameters.
 * @param options - The options for the setRoyaltyEngine function.
 * @returns The encoded hexadecimal string.
 * @extension COMMON
 * @example
 * ```ts
 * import { encodeSetRoyaltyEngine } "thirdweb/extensions/common";
 * const result = encodeSetRoyaltyEngine({
 *  royaltyEngineAddress: ...,
 * });
 * ```
 */
export function encodeSetRoyaltyEngine(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeSetRoyaltyEngineParams(options).slice(2));
}
/**
 * Prepares a transaction to call the "setRoyaltyEngine" function on the contract.
 * @param options - The options for the "setRoyaltyEngine" function.
 * @returns A prepared transaction object.
 * @extension COMMON
 * @example
 * ```ts
 * import { setRoyaltyEngine } from "thirdweb/extensions/common";
 *
 * const transaction = setRoyaltyEngine({
 *  contract,
 *  royaltyEngineAddress: ...,
 *  overrides: {
 *    ...
 *  }
 * });
 *
 * // Send the transaction
 * ...
 *
 * ```
 */
export function setRoyaltyEngine(options) {
    const asyncOptions = once(async () => {
        return "asyncParams" in options ? await options.asyncParams() : options;
    });
    return prepareContractCall({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: async () => {
            const resolvedOptions = await asyncOptions();
            return [resolvedOptions.royaltyEngineAddress];
        },
        value: async () => (await asyncOptions()).overrides?.value,
        accessList: async () => (await asyncOptions()).overrides?.accessList,
        gas: async () => (await asyncOptions()).overrides?.gas,
        gasPrice: async () => (await asyncOptions()).overrides?.gasPrice,
        maxFeePerGas: async () => (await asyncOptions()).overrides?.maxFeePerGas,
        maxPriorityFeePerGas: async () => (await asyncOptions()).overrides?.maxPriorityFeePerGas,
        nonce: async () => (await asyncOptions()).overrides?.nonce,
        extraGas: async () => (await asyncOptions()).overrides?.extraGas,
    });
}
