// thirdweb/dist/cjs/react/web/wallets/in-app/InAppWalletPassKeyLogin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InAppWalletPassKeyLogin = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const passkeys_js_1 = require("../../../../wallets/in-app/implementations/lib/auth/passkeys.js");
const useWalletConnectionCtx_js_1 = require("../../../core/hooks/others/useWalletConnectionCtx.js");
const AccentFailIcon_js_1 = require("../../ui/ConnectWallet/icons/AccentFailIcon.js");
const FingerPrintIcon_js_1 = require("../../ui/ConnectWallet/icons/FingerPrintIcon.js");
const Spacer_js_1 = require("../../ui/components/Spacer.js");
const Spinner_js_1 = require("../../ui/components/Spinner.js");
const basic_js_1 = require("../../ui/components/basic.js");
const buttons_js_1 = require("../../ui/components/buttons.js");
const text_js_1 = require("../../ui/components/text.js");
const index_js_1 = require("../../ui/design-system/index.js");
const LoadingScreen_js_1 = require("../shared/LoadingScreen.js");
const storage_js_1 = require("./storage.js");
// is passkey stored?
// - login
// else
// - show login or signup options
function InAppWalletPassKeyLogin(props) {
    const { client, connectModal, chain } = (0, useWalletConnectionCtx_js_1.useConnectUI)();
    const { wallet, done } = props;
    const [screen, setScreen] = (0, react_1.useState)("loading");
    const triggered = (0, react_1.useRef)(false);
    (0, react_1.useEffect)(() => {
        if (triggered.current) {
            return;
        }
        triggered.current = true;
        (0, passkeys_js_1.hasStoredPasskey)(client)
            .then((isStored) => {
            if (isStored) {
                setScreen("login");
            }
            else {
                setScreen("select");
            }
        })
            .catch(() => {
            setScreen("select");
        });
    }, [client]);
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { animate: "fadein", fullHeight: true, flex: "column", children: [(0, jsx_runtime_1.jsx)(basic_js_1.Container, { p: "lg", children: (0, jsx_runtime_1.jsx)(basic_js_1.ModalHeader, { title: "Passkey", onBack: props.onBack }) }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { px: connectModal.size === "wide" ? "xxl" : "lg", expand: true, flex: "column", center: "y", children: (0, jsx_runtime_1.jsxs)("div", { children: [screen === "loading" && ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(LoadingScreen_js_1.LoadingScreen, {}), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xxl" })] })), screen === "select" && ((0, jsx_runtime_1.jsx)(SelectLoginMethod, { onSignin: () => {
                                setScreen("login");
                            }, onSignup: () => {
                                setScreen("signup");
                            } })), screen === "login" && ((0, jsx_runtime_1.jsx)(LoginScreen, { wallet: wallet, client: client, done: done, onCreate: () => {
                                setScreen("signup");
                            }, chain: chain })), screen === "signup" && ((0, jsx_runtime_1.jsx)(SignupScreen, { wallet: wallet, client: client, done: done, chain: chain }))] }) })] }));
}
exports.InAppWalletPassKeyLogin = InAppWalletPassKeyLogin;
function LoginScreen(props) {
    const { wallet, done, client, chain } = props;
    const [status, setStatus] = (0, react_1.useState)("loading");
    async function login() {
        setStatus("loading");
        try {
            await wallet.connect({
                client: client,
                strategy: "passkey",
                type: "sign-in",
                chain,
            });
            await (0, storage_js_1.setLastAuthProvider)("passkey");
            done();
        }
        catch {
            setStatus("error");
        }
    }
    const triggered = (0, react_1.useRef)(false);
    (0, react_1.useEffect)(() => {
        if (triggered.current) {
            return;
        }
        triggered.current = true;
        login();
    });
    if (status === "loading") {
        return ((0, jsx_runtime_1.jsx)(LoadingState, { title: "Requesting Passkey", subtitle: "A pop-up prompt will appear to sign-in and verify your passkey" }));
    }
    if (status === "error") {
        return ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(ErrorState, { onTryAgain: login, title: "Failed to Login" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" }), (0, jsx_runtime_1.jsx)(buttons_js_1.Button, { variant: "outline", fullWidth: true, onClick: props.onCreate, children: "Create a new Passkey" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" })] }));
    }
    return null;
}
function SignupScreen(props) {
    const { wallet, done, client, chain } = props;
    const [status, setStatus] = (0, react_1.useState)("loading");
    async function signup() {
        setStatus("loading");
        try {
            await wallet.connect({
                client: client,
                strategy: "passkey",
                type: "sign-up",
                chain,
            });
            await (0, storage_js_1.setLastAuthProvider)("passkey");
            done();
        }
        catch {
            setStatus("error");
        }
    }
    const triggered = (0, react_1.useRef)(false);
    (0, react_1.useEffect)(() => {
        if (triggered.current) {
            return;
        }
        triggered.current = true;
        signup();
    });
    if (status === "loading") {
        return ((0, jsx_runtime_1.jsx)(LoadingState, { title: "Creating Passkey", subtitle: "A pop-up prompt will appear to sign-in and verify your passkey" }));
    }
    if (status === "error") {
        return ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: [(0, jsx_runtime_1.jsx)(ErrorState, { onTryAgain: signup, title: "Failed to create passkey" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" })] }));
    }
    return null;
}
function SelectLoginMethod(props) {
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xxl" }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", center: "x", color: "accentText", children: (0, jsx_runtime_1.jsx)(FingerPrintIcon_js_1.FingerPrintIcon, { size: index_js_1.iconSize["4xl"] }) }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xl" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xxl" }), (0, jsx_runtime_1.jsx)(buttons_js_1.Button, { variant: "accent", onClick: props.onSignup, fullWidth: true, children: "Create a Passkey" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "sm" }), (0, jsx_runtime_1.jsx)(buttons_js_1.Button, { variant: "outline", onClick: props.onSignin, fullWidth: true, children: "I have a Passkey" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" })] }));
}
function ErrorState(props) {
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { animate: "fadein", children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xxl" }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { flex: "row", center: "x", children: (0, jsx_runtime_1.jsx)(AccentFailIcon_js_1.AccentFailIcon, { size: index_js_1.iconSize["3xl"] }) }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "lg" }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { center: true, color: "primaryText", size: "lg", children: props.title }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xl" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xxl" }), (0, jsx_runtime_1.jsx)(buttons_js_1.Button, { variant: "accent", fullWidth: true, onClick: props.onTryAgain, children: "Try Again" })] }));
}
function LoadingState(props) {
    return ((0, jsx_runtime_1.jsxs)(basic_js_1.Container, { animate: "fadein", children: [(0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xxl" }), (0, jsx_runtime_1.jsxs)(basic_js_1.Container, { flex: "row", center: "x", style: {
                    position: "relative",
                }, children: [(0, jsx_runtime_1.jsx)(Spinner_js_1.Spinner, { size: "4xl", color: "accentText" }), (0, jsx_runtime_1.jsx)(basic_js_1.Container, { color: "accentText", style: {
                            position: "absolute",
                            top: "50%",
                            left: "50%",
                            transform: "translate(-50%, -50%)",
                        }, children: (0, jsx_runtime_1.jsx)(FingerPrintIcon_js_1.FingerPrintIcon, { size: index_js_1.iconSize.xxl }) })] }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xl" }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { center: true, color: "primaryText", size: "lg", children: props.title }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "md" }), (0, jsx_runtime_1.jsx)(text_js_1.Text, { multiline: true, center: true, children: props.subtitle }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xxl" }), (0, jsx_runtime_1.jsx)(Spacer_js_1.Spacer, { y: "xxl" })] }));
}
